// @sanity/import/dist/importBatches.js
import partition from 'lodash-es/partition.js';
import pMap from 'p-map';
import { progressStepper } from './util/progressStepper.js';
import { retryOnFailure } from './util/retryOnFailure.js';
import { suffixTag } from './util/suffixTag.js';
const DOCUMENT_IMPORT_CONCURRENCY = 6;
export async function importBatches(batches, options) {
    const progress = progressStepper(options.onProgress, {
        step: 'Importing documents',
        total: batches.length
    });
    const mapOptions = {
        concurrency: DOCUMENT_IMPORT_CONCURRENCY
    };
    const batchResults = await pMap(batches, importBatch.bind(null, options, progress), mapOptions);
    const result = {
        count: 0,
        importedIds: []
    };
    for (const batchResult of batchResults){
        result.count += batchResult.count;
        result.importedIds = [
            ...result.importedIds,
            ...batchResult.importedIds
        ];
    }
    return result;
}
function importBatch(options, progress, batch) {
    const { client, operation, releasesOperation, tag } = options;
    const maxRetries = operation === 'create' ? 1 : 3;
    return retryOnFailure(()=>{
        const [releaseDocs, docs] = partition(batch, (doc)=>doc._id.startsWith('_.releases.'));
        const docsTransaction = docs.length > 0 ? (()=>{
            let trx = client.transaction();
            for (const doc of docs){
                switch(operation){
                    case 'create':
                        {
                            trx = trx.create(doc);
                            break;
                        }
                    case 'createIfNotExists':
                        {
                            trx = trx.createIfNotExists(doc);
                            break;
                        }
                    case 'createOrReplace':
                        {
                            trx = trx.createOrReplace(doc);
                            break;
                        }
                    default:
                        {
                            throw new Error(`Unknown operation: ${operation}`);
                        }
                }
            }
            return trx;
        })().commit({
            tag: suffixTag(tag, 'doc.create'),
            visibility: 'async'
        }).then((res)=>{
            progress();
            const importedIds = res.results.filter((r)=>r.operation !== 'none').map((r)=>r.id);
            return {
                count: res.results.length,
                importedIds
            };
        }) : Promise.resolve({
            count: 0,
            importedIds: []
        });
        const releasesAction = releaseDocs.map((doc)=>{
            const actionParams = {
                actionType: 'sanity.action.release.import',
                attributes: doc,
                ifExists: releasesOperation,
                releaseId: doc.name
            };
            return client.action(actionParams).then(()=>({
                    count: 1,
                    importedIds: [
                        doc._id
                    ]
                })).catch((err)=>{
                err.message = `Release import failed for ${doc._id}: ${err.message}`;
                throw err;
            });
        });
        return Promise.all([
            docsTransaction,
            ...releasesAction
        ]).then((results)=>{
            const combined = {
                count: 0,
                importedIds: []
            };
            for (const r of results){
                combined.count += r.count;
                combined.importedIds = [
                    ...combined.importedIds,
                    ...r.importedIds
                ];
            }
            return combined;
        });
    }, {
        isRetriable,
        maxTries: maxRetries
    });
}
function isRetriable(err) {
    return !err.response || err.response.statusCode !== 409;
}
