// @sanity/cli/dist/services/auth.js
import { getGlobalCliClient } from '@sanity/cli-core';
export const AUTH_API_VERSION = 'v2025-09-23';
async function getUnauthenticatedClient() {
    return getGlobalCliClient({
        apiVersion: AUTH_API_VERSION,
        unauthenticated: true
    });
}
/**
 * Invalidate the current user session
 *
 * @param token - Optional token to invalidate.
 */ export async function logout(token) {
    let client = await getGlobalCliClient({
        apiVersion: AUTH_API_VERSION
    });
    if (token) {
        client = client.withConfig({
            token
        });
    }
    return client.request({
        method: 'POST',
        uri: '/auth/logout'
    });
}
export async function getProviders() {
    const client = await getUnauthenticatedClient();
    return client.request({
        uri: '/auth/providers'
    });
}
export async function getVercelProviderUrl() {
    const client = (await getUnauthenticatedClient()).withConfig({
        apiVersion: 'v1'
    });
    return client.getUrl('/auth/login/vercel');
}
export async function getSSOProviders(orgSlug) {
    const client = await getUnauthenticatedClient();
    return client.request({
        uri: `/auth/organizations/by-slug/${orgSlug}/providers`
    });
}
export async function getTokenDetails(queryString) {
    const client = await getUnauthenticatedClient();
    return client.request({
        uri: `/auth/fetch${queryString}`
    });
}
