// @sanity/cli/dist/services/backup.js
import { getGlobalCliClient } from '@sanity/cli-core';
import { BACKUP_API_VERSION } from '../actions/backup/constants.js';
export async function listBackups(options) {
    const client = await getGlobalCliClient({
        apiVersion: BACKUP_API_VERSION,
        requireUser: true
    });
    const query = {};
    if (options.limit) {
        query.limit = options.limit.toString();
    }
    if (options.after) {
        query.after = options.after;
    }
    if (options.before) {
        query.before = options.before;
    }
    return client.request({
        query,
        uri: `/projects/${options.projectId}/datasets/${options.datasetName}/backups`
    });
}
export async function getBackupDetails(options) {
    const client = await getGlobalCliClient({
        apiVersion: BACKUP_API_VERSION,
        requireUser: true
    });
    const query = {};
    if (options.nextCursor) {
        query.nextCursor = options.nextCursor;
    }
    return client.request({
        query,
        uri: `/projects/${options.projectId}/datasets/${options.datasetName}/backups/${options.backupId}`
    });
}
export async function setBackup({ dataset, projectId, status }) {
    const client = await getGlobalCliClient({
        apiVersion: BACKUP_API_VERSION,
        requireUser: true
    });
    return client.request({
        body: {
            enabled: status
        },
        method: 'PUT',
        uri: `/projects/${projectId}/datasets/${dataset}/settings/backups`
    });
}
