// @sanity/cli/dist/services/datasets.js
import { getProjectCliClient } from '@sanity/cli-core';
import { EventSource } from 'eventsource';
import { Observable } from 'rxjs';
export const DATASET_API_VERSION = 'v2025-09-16';
function getDatasetClient(projectId) {
    return getProjectCliClient({
        apiVersion: DATASET_API_VERSION,
        projectId,
        requireUser: true
    });
}
export async function listDatasets(projectId) {
    const client = await getDatasetClient(projectId);
    return client.datasets.list();
}
export async function listDatasetAliases(projectId) {
    const client = await getDatasetClient(projectId);
    return client.request({
        uri: '/aliases'
    });
}
export async function deleteDataset({ datasetName, projectId }) {
    const client = await getDatasetClient(projectId);
    return client.datasets.delete(datasetName);
}
export async function editDatasetAcl({ aclMode, datasetName, projectId }) {
    const client = await getDatasetClient(projectId);
    return client.datasets.edit(datasetName, {
        aclMode
    });
}
export async function createDataset({ aclMode, datasetName, embeddings, projectId }) {
    const client = await getDatasetClient(projectId);
    const options = {};
    if (aclMode) {
        options.aclMode = aclMode;
    }
    if (embeddings) {
        options.embeddings = embeddings;
    }
    return client.datasets.create(datasetName, options);
}
export async function copyDataset({ projectId, skipHistory, sourceDataset, targetDataset }) {
    const client = await getDatasetClient(projectId);
    return client.request({
        body: {
            skipHistory,
            targetDataset
        },
        method: 'PUT',
        uri: `/datasets/${sourceDataset}/copy`
    });
}
export async function listDatasetCopyJobs({ limit, offset, projectId }) {
    const client = await getDatasetClient(projectId);
    const query = {};
    if (offset !== undefined && offset >= 0) {
        query.offset = `${offset}`;
    }
    if (limit !== undefined && limit > 0) {
        query.limit = `${limit}`;
    }
    return client.request({
        method: 'GET',
        query,
        uri: `/projects/${projectId}/datasets/copy`
    });
}
async function getJobListenUrl(projectId, jobId) {
    const client = await getDatasetClient(projectId);
    const baseUrl = client.config().url || 'https://api.sanity.io';
    return `${baseUrl}/jobs/${jobId}/listen`;
}
export function followCopyJobProgress({ jobId, projectId }) {
    return new Observable((observer)=>{
        let progressSource = null;
        let stopped = false;
        getJobListenUrl(projectId, jobId).then((url)=>{
            progressSource = new EventSource(url);
            function onError() {
                if (progressSource) {
                    progressSource.close();
                    progressSource = null;
                }
                if (stopped) {
                    return;
                }
                observer.next({
                    type: 'reconnect'
                });
                progressSource = new EventSource(url);
                attachListeners();
            }
            function onChannelError(error) {
                stopped = true;
                if (progressSource) {
                    progressSource.close();
                    progressSource = null;
                }
                const errorMessage = error.data ? `Copy job failed: ${error.data}` : 'Copy job failed: Connection to server lost. Please check the job status using --list and retry if needed.';
                observer.error(new Error(errorMessage));
            }
            function onMessage(event) {
                let data;
                try {
                    data = JSON.parse(event.data);
                } catch (error) {
                    const message = error instanceof Error ? error.message : 'Unknown error';
                    observer.error(new Error(`Invalid JSON received from server: ${message}`));
                    return;
                }
                if (data.state === 'failed') {
                    const failureReason = data.message || data.error || 'Unknown reason';
                    observer.error(new Error(`Copy job failed: ${failureReason}`));
                } else if (data.state === 'completed') {
                    onComplete();
                } else {
                    observer.next(data);
                }
            }
            function onComplete() {
                if (progressSource) {
                    progressSource.removeEventListener('error', onError);
                    progressSource.removeEventListener('channel_error', onChannelError);
                    progressSource.removeEventListener('job', onMessage);
                    progressSource.removeEventListener('done', onComplete);
                    progressSource.close();
                    progressSource = null;
                }
                observer.complete();
            }
            function attachListeners() {
                if (progressSource) {
                    progressSource.addEventListener('error', onError);
                    progressSource.addEventListener('channel_error', onChannelError);
                    progressSource.addEventListener('job', onMessage);
                    progressSource.addEventListener('done', onComplete);
                }
            }
            attachListeners();
        }).catch((error)=>{
            observer.error(error);
        });
        return ()=>{
            if (stopped) return;
            stopped = true;
            if (progressSource) {
                progressSource.close();
                progressSource = null;
            }
        };
    });
}
