// @sanity/cli/dist/services/documents.js
import { getProjectCliClient } from '@sanity/cli-core';
export const DOCUMENTS_API_VERSION = 'v2021-03-25';
function getDocumentsClient({ dataset, projectId }) {
    return getProjectCliClient({
        apiVersion: DOCUMENTS_API_VERSION,
        dataset,
        projectId,
        requireUser: true
    });
}
/**
 * Get the total count of documents in the dataset
 */ export async function getDocumentCount({ dataset, projectId }) {
    const client = await getDocumentsClient({
        dataset,
        projectId
    });
    return client.fetch('length(*)');
}
/**
 * Export all documents from the dataset as a streaming response
 */ export async function exportDocuments({ dataset, projectId }) {
    const client = await getDocumentsClient({
        dataset,
        projectId
    });
    const exportUrl = new URL(client.getUrl(`/data/export/${dataset}`, false));
    const { token } = client.config();
    const response = await fetch(exportUrl, {
        // eslint-disable-next-line n/no-unsupported-features/node-builtins -- Headers is stable in modern Node.js
        headers: new Headers({
            ...token && {
                Authorization: `Bearer ${token}`
            }
        })
    });
    return response;
}
/**
 * Check the availability/existence of documents by their IDs
 */ export async function checkDocumentAvailability({ dataset, documentIds, projectId }) {
    const client = await getDocumentsClient({
        dataset,
        projectId
    });
    return client.request({
        json: true,
        query: {
            excludeContent: 'true'
        },
        tag: 'documents-availability',
        uri: client.getDataUrl('doc', documentIds.join(','))
    });
}
