// @sanity/cli/dist/services/graphql.js
import { getProjectCliClient } from '@sanity/cli-core';
import { getUrlHeaders } from './getUrlHeaders.js';
export const GRAPHQL_API_VERSION = 'v2025-09-19';
/**
 * List all GraphQL endpoints for a project
 * @param client - The API client to use for the request
 * @returns A promise that resolves to an array of GraphQL endpoints
 *
 * @internal
 */ export async function listGraphQLEndpoints(projectId) {
    const client = await getProjectCliClient({
        apiVersion: GRAPHQL_API_VERSION,
        projectId,
        requireUser: true
    });
    return client.request({
        method: 'GET',
        uri: '/apis/graphql'
    });
}
export async function deleteGraphQLAPI({ dataset, projectId, tag }) {
    const client = await getProjectCliClient({
        apiVersion: GRAPHQL_API_VERSION,
        projectId,
        requireUser: true
    });
    return client.request({
        method: 'DELETE',
        uri: `/apis/graphql/${dataset}/${tag}`
    });
}
export async function validateGraphQLAPI({ dataset, enablePlayground, projectId, schema, tag }) {
    const client = await getProjectCliClient({
        apiVersion: GRAPHQL_API_VERSION,
        projectId,
        requireUser: true
    });
    return client.request({
        body: {
            enablePlayground,
            schema
        },
        maxRedirects: 0,
        method: 'POST',
        url: `/apis/graphql/${dataset}/${tag}/validate`
    });
}
export async function deployGraphQLAPI({ dataset, enablePlayground, projectId, schema, tag }) {
    const client = await getProjectCliClient({
        apiVersion: GRAPHQL_API_VERSION,
        projectId,
        requireUser: true
    });
    return client.request({
        body: {
            enablePlayground,
            schema
        },
        maxRedirects: 0,
        method: 'PUT',
        url: `/apis/graphql/${dataset}/${tag}`
    });
}
export async function getClientUrl(projectId, uri) {
    const client = await getProjectCliClient({
        apiVersion: GRAPHQL_API_VERSION,
        projectId,
        requireUser: true
    });
    return `${client.config().url}/${uri.replace(/^\//, '')}`;
}
export async function getCurrentSchemaProps(projectId, dataset, tag) {
    try {
        const client = await getProjectCliClient({
            apiVersion: GRAPHQL_API_VERSION,
            projectId
        });
        const uri = `/apis/graphql/${dataset}/${tag}`;
        const config = client.config();
        const apiUrl = `${config.url}/${uri.replace(/^\//, '')}`;
        const res = await getUrlHeaders(apiUrl, {
            Authorization: `Bearer ${config.token}`
        });
        return {
            currentGeneration: res['x-sanity-graphql-generation'],
            playgroundEnabled: res['x-sanity-graphql-playground'] === 'true'
        };
    } catch (err) {
        if (err instanceof Error && 'response' in err && typeof err.response === 'object' && err.response !== null && 'statusCode' in err.response && err.response.statusCode === 404) {
            return {};
        }
        throw err;
    }
}
