// @sanity/cli/dist/services/hooks.js
import { getGlobalCliClient } from '@sanity/cli-core';
import { HOOK_API_VERSION } from '../actions/hook/constants.js';
export async function getHooksForProject(projectId) {
    const client = await getGlobalCliClient({
        apiVersion: HOOK_API_VERSION,
        requireUser: true
    });
    return client.request({
        uri: `/hooks/projects/${projectId}`
    });
}
export async function getHookMessagesForProject({ hookId, projectId }) {
    const client = await getGlobalCliClient({
        apiVersion: HOOK_API_VERSION,
        requireUser: true
    });
    return client.request({
        uri: `/hooks/projects/${projectId}/${hookId}/messages`
    });
}
export async function getHookAttemptsForProject({ hookId, projectId }) {
    const client = await getGlobalCliClient({
        apiVersion: HOOK_API_VERSION,
        requireUser: true
    });
    return client.request({
        uri: `/hooks/projects/${projectId}/${hookId}/attempts`
    });
}
export async function getHookAttempt({ attemptId, projectId }) {
    const client = await getGlobalCliClient({
        apiVersion: HOOK_API_VERSION,
        requireUser: true
    });
    return client.request({
        uri: `/hooks/projects/${projectId}/attempts/${attemptId}`
    });
}
export async function listHooksForProject(projectId) {
    const client = await getGlobalCliClient({
        apiVersion: HOOK_API_VERSION,
        requireUser: true
    });
    return client.request({
        uri: `/hooks/projects/${projectId}`
    });
}
export async function deleteHookForProject(projectId, hookId) {
    const client = await getGlobalCliClient({
        apiVersion: HOOK_API_VERSION,
        requireUser: true
    });
    return client.request({
        method: 'DELETE',
        uri: `/hooks/projects/${projectId}/${hookId}`
    });
}
