// @sanity/cli/dist/services/schemas.js
import { getGlobalCliClient } from '@sanity/cli-core';
export const SCHEMA_API_VERSION = 'v2025-03-01';
async function getSchemaClient() {
    return await getGlobalCliClient({
        apiVersion: SCHEMA_API_VERSION,
        requireUser: true
    });
}
export async function getSchemas(dataset, projectId, id) {
    const client = await getSchemaClient();
    return client.request({
        method: 'GET',
        uri: `/projects/${projectId}/datasets/${dataset}/schemas${id ? `/${id}` : ''}`
    });
}
export async function deleteSchema(dataset, projectId, id) {
    const exists = await getSchemas(dataset, projectId, id);
    if (exists?.length === 0) {
        return {
            deleted: false
        };
    }
    const client = await getSchemaClient();
    return client.request({
        method: 'DELETE',
        uri: `/projects/${projectId}/datasets/${dataset}/schemas/${id}`
    });
}
export async function updateSchemas(dataset, projectId, schemas) {
    const client = await getSchemaClient();
    return client.request({
        body: {
            schemas
        },
        method: 'PUT',
        url: `/projects/${projectId}/datasets/${dataset}/schemas`
    });
}
