// @sanity/cli/dist/actions/debug/gatherDebugInfo.js
import { getCliToken, getUserConfig } from '@sanity/cli-core';
import { getProjectById } from '../../services/projects.js';
import { getCliUser, getProjectUser } from '../../services/user.js';
import { findSanityModulesVersions } from '../versions/findSanityModulesVersions.js';
export async function gatherDebugInfo(options) {
    const { cliConfig, includeSecrets, projectRoot } = options;
    // Gather all info in parallel where possible
    const [auth, globalConfig, projectConfigResult, versions] = await Promise.all([
        gatherAuthInfo(includeSecrets),
        gatherGlobalConfig(),
        cliConfig ? gatherProjectConfig(cliConfig) : undefined,
        gatherVersionsInfo(projectRoot)
    ]);
    // Gather user and project info that depend on auth
    const user = await gatherUserInfo(projectConfigResult, auth.hasToken);
    const project = await gatherProjectInfo(projectConfigResult, auth.hasToken, user);
    return {
        auth,
        globalConfig,
        project,
        projectConfig: projectConfigResult,
        user,
        versions
    };
}
async function gatherAuthInfo(includeSecrets) {
    const token = await getCliToken();
    const hasToken = Boolean(token);
    return {
        authToken: includeSecrets && token ? token : '<redacted>',
        hasToken,
        userType: 'normal'
    };
}
function gatherGlobalConfig() {
    return getUserConfig().all;
}
function gatherProjectConfig(cliConfig) {
    if (!cliConfig.api?.projectId) {
        return new Error('Missing required "api.projectId" key');
    }
    return cliConfig;
}
async function gatherVersionsInfo(projectRoot) {
    if (!projectRoot) {
        return undefined;
    }
    try {
        return await findSanityModulesVersions({
            cwd: projectRoot.directory
        });
    } catch  {
        return [];
    }
}
async function gatherUserInfo(projectConfig, hasToken) {
    if (!hasToken) {
        return new Error('Not logged in');
    }
    try {
        /**
     * If the project config has a project ID, get the user for the project
     * Otherwise, get the user for the global client
     */ const userInfo = !projectConfig || projectConfig instanceof Error || !projectConfig.api?.projectId ? await getCliUser() : await getProjectUser(projectConfig.api.projectId);
        return {
            email: userInfo.email,
            id: userInfo.id,
            name: userInfo.name
        };
    } catch (error) {
        return error instanceof Error ? error : new Error('Failed to fetch user info');
    }
}
async function gatherProjectInfo(projectConfig, hasToken, user) {
    if (!hasToken || !projectConfig || projectConfig instanceof Error) {
        return null;
    }
    const projectId = projectConfig.api?.projectId;
    if (!projectId) {
        return null;
    }
    try {
        const projectInfo = await getProjectById(projectId);
        if (!projectInfo) {
            return new Error(`Project specified in configuration (${projectId}) does not exist in API`);
        }
        const userId = user instanceof Error || !user ? null : user.id;
        const member = (projectInfo.members || []).find((member)=>member.id === userId);
        return {
            displayName: projectInfo.displayName,
            id: projectId,
            // @ts-expect-error - Incorrect type definition in @sanity/client
            userRoles: member && member.roles ? member.roles.map((role)=>role.name) : [
                '<none>'
            ]
        };
    } catch (error) {
        return error instanceof Error ? error : new Error('Failed to fetch project info');
    }
}
