// @sanity/cli/dist/commands/telemetry/enable.js
import { SanityCommand } from '@sanity/cli-core';
import { setConsent } from '../../actions/telemetry/setConsent.js';
import { telemetryLearnMoreMessage } from '../../actions/telemetry/telemetryLearnMoreMessage.js';
export class Enable extends SanityCommand {
    static description = 'Enable telemetry for your logged in user';
    static examples = [
        {
            command: '<%= config.bin %> telemetry <%= command.id %>',
            description: 'Enable telemetry for your logged in user'
        }
    ];
    static flags = {};
    async run() {
        // Parse to ensure no invalid flags are passed
        await this.parse(Enable);
        try {
            const result = await setConsent({
                status: 'granted'
            });
            this.log(result.message);
            if (result.changed) {
                this.log(`\n${telemetryLearnMoreMessage('granted')}`);
            }
        } catch (err) {
            const message = err instanceof Error ? err.message : 'An unknown error occurred';
            this.error(message, {
                exit: 1
            });
        }
    }
}
