// @sanity/cli/dist/actions/schema/getExtractOptions.js
import { existsSync, statSync } from 'node:fs';
import { extname, join, resolve } from 'node:path';
export function getExtractOptions({ flags, projectRoot, schemaExtraction }) {
    const pathFlag = flags.path ?? schemaExtraction?.path;
    let outputPath;
    if (pathFlag) {
        const resolved = resolve(join(projectRoot.directory, pathFlag));
        const isExistingDirectory = existsSync(resolved) && statSync(resolved).isDirectory();
        outputPath = isExistingDirectory || !extname(resolved) ? join(resolved, 'schema.json') : resolved;
    } else {
        outputPath = resolve(join(projectRoot.directory, 'schema.json'));
    }
    return {
        configPath: projectRoot.path,
        enforceRequiredFields: flags['enforce-required-fields'] ?? schemaExtraction?.enforceRequiredFields ?? false,
        format: flags.format ?? 'groq-type-nodes',
        outputPath,
        watchPatterns: flags['watch-patterns'] ?? schemaExtraction?.watchPatterns ?? [],
        workspace: flags.workspace ?? schemaExtraction?.workspace
    };
}
