// @sanity/cli/dist/commands/cors/list.js
import { SanityCommand, subdebug } from '@sanity/cli-core';
import { promptForProject } from '../../prompts/promptForProject.js';
import { listCorsOrigins } from '../../services/cors.js';
import { getProjectIdFlag } from '../../util/sharedFlags.js';
const listCorsDebug = subdebug('cors:list');
export class List extends SanityCommand {
    static description = 'List all origins allowed to access the API for this project';
    static examples = [
        {
            command: '<%= config.bin %> <%= command.id %>',
            description: 'List CORS origins for the current project'
        },
        {
            command: '<%= config.bin %> <%= command.id %> --project-id abc123',
            description: 'List CORS origins for a specific project'
        }
    ];
    static flags = {
        ...getProjectIdFlag({
            description: 'Project ID to list CORS origins for',
            semantics: 'override'
        })
    };
    async run() {
        await this.parse(List);
        const projectId = await this.getProjectId({
            fallback: ()=>promptForProject({
                    requiredPermissions: [
                        {
                            grant: 'read',
                            permission: 'sanity.project.cors'
                        }
                    ]
                })
        });
        let origins;
        try {
            origins = await listCorsOrigins(projectId);
        } catch (error) {
            const err = error;
            listCorsDebug(`Error fetching CORS origins for project ${projectId}`, err);
            this.error(`CORS origins list retrieval failed:\n${err.message}`, {
                exit: 1
            });
        }
        if (origins.length === 0) {
            this.log('No CORS origins configured for this project.');
            return;
        }
        // Output each origin on a new line, matching the original behavior
        this.log(origins.map((origin)=>origin.origin).join('\n'));
    }
}
