// @sanity/cli-core/dist/util/getCliTelemetry.js
import { ux } from '@oclif/core';
import { noopLogger } from '@sanity/telemetry';
/**
 * @public
 * Symbol used to store CLI telemetry state on globalThis.
 * Use the accessor functions instead of accessing this directly.
 */ export const CLI_TELEMETRY_SYMBOL = Symbol.for('sanity.cli.telemetry');
function getState() {
    return globalThis[CLI_TELEMETRY_SYMBOL];
}
/**
 * @public
 */ export function getCliTelemetry() {
    const state = getState();
    // This should never happen, but if it does, we return a noop logger to avoid errors.
    if (!state) {
        ux.warn('CLI telemetry not initialized, returning noop logger');
        return noopLogger;
    }
    return state.logger;
}
/**
 * Sets the global CLI telemetry state.
 * @internal
 */ export function setCliTelemetry(telemetry, options) {
    ;
    globalThis[CLI_TELEMETRY_SYMBOL] = {
        logger: telemetry,
        reportTraceError: options?.reportTraceError
    };
}
/**
 * Reports an error to the CLI command trace. Called from SanityCommand.catch()
 * for real command errors (not user aborts).
 * @internal
 */ export function reportCliTraceError(error) {
    getState()?.reportTraceError?.(error);
}
/**
 * Clears the global CLI telemetry store.
 * @internal
 */ export function clearCliTelemetry() {
    const global = globalThis;
    delete global[CLI_TELEMETRY_SYMBOL];
}
