// @sanity/cli/dist/actions/documents/validation/reporters/prettyReporter/prettyReporter.js
import { spinner } from '@sanity/cli-core/ux';
import { formatDocumentValidation } from './formatDocumentValidation.js';
import { count, levelValues, percent, seconds, summary } from './util.js';
/**
 * Represents the default stylish/pretty reporter
 */ export const pretty = async ({ flags, output, worker })=>{
    const workspaceLoadStart = Date.now();
    // Report workspace loaded
    const spin = spinner(flags.workspace ? `Loading workspace '${flags.workspace}'…` : 'Loading workspace…').start();
    const workspace = await worker.event.loadedWorkspace();
    spin.succeed(`Loaded workspace '${workspace.name}' using project '${workspace.projectId}' and dataset '${flags.dataset || workspace.dataset}' ${seconds(workspaceLoadStart)}`);
    if (!flags.file) {
        // Report document count
        spin.start('Calculating documents to be validated…');
        const { documentCount } = await worker.event.loadedDocumentCount();
        // Report export progress
        const downloadStart = Date.now();
        spin.text = `Downloading ${count(documentCount, 'documents')}…`;
        for await (const { downloadedCount } of worker.stream.exportProgress()){
            const percentage = percent(downloadedCount / documentCount);
            spin.text = `Downloading ${count(documentCount, 'documents')}… ${percentage}`;
        }
        spin.succeed(`Downloaded ${count(documentCount, 'documents')} ${seconds(downloadStart)}`);
    }
    const { totalDocumentsToValidate } = await worker.event.exportFinished();
    const referenceIntegrityStart = Date.now();
    spin.start(`Checking reference existence…`);
    await worker.event.loadedReferenceIntegrity();
    spin.succeed(`Checked all references ${seconds(referenceIntegrityStart)}`);
    // Report validation progress
    const validationStart = Date.now();
    spin.start(`Validating ${count(totalDocumentsToValidate, 'documents')}…`);
    const results = [];
    const totals = {
        errors: {
            documents: 0,
            markers: 0
        },
        infos: {
            documents: 0,
            markers: 0
        },
        valid: {
            documents: 0
        },
        warnings: {
            documents: 0,
            markers: 0
        }
    };
    for await (const { validatedCount, ...result } of worker.stream.validation()){
        const { markers } = result;
        if (markers.length > 0) {
            results.push(result);
        }
        const errors = markers.filter((marker)=>marker.level === 'error');
        const warnings = markers.filter((marker)=>marker.level === 'warning');
        const infos = markers.filter((marker)=>marker.level === 'info');
        if (markers.length === 0) {
            totals.valid.documents += 1;
        }
        if (errors.length > 0) {
            totals.errors.documents += 1;
            totals.errors.markers += errors.length;
        }
        if (warnings.length > 0) {
            totals.warnings.documents += 1;
            totals.warnings.markers += warnings.length;
        }
        if (infos.length > 0) {
            totals.infos.documents += 1;
            totals.infos.markers += infos.length;
        }
        spin.text = `Validating ${count(totalDocumentsToValidate, 'documents')}…\n\n` + `Processed ${count(validatedCount, 'documents')} (${percent(validatedCount / totalDocumentsToValidate)}):\n${summary(totals, flags.level)}`;
    }
    spin.succeed(`Validated ${count(totalDocumentsToValidate, 'documents')} ${seconds(validationStart)}`);
    output.log(`\nValidation results:\n${summary(totals, flags.level)}`);
    results.sort((a, b)=>{
        if (a.level === b.level) return a.documentType.localeCompare(b.documentType);
        return levelValues[a.level] - levelValues[b.level];
    });
    let overallLevel = 'info';
    for (const result of results){
        if (result.level === 'error') overallLevel = 'error';
        if (result.level === 'warning' && overallLevel !== 'error') overallLevel = 'warning';
        output.log(`${formatDocumentValidation(result)}\n`);
    }
    await worker.dispose();
    return overallLevel;
};
