// @sanity/cli/dist/actions/build/buildStaticFiles.js
import path from 'node:path';
import { build } from 'vite';
import { copyDir } from '../../util/copyDir.js';
import { buildDebug } from './buildDebug.js';
import { extendViteConfigWithUserConfig, finalizeViteConfig, getViteConfig } from './getViteConfig.js';
import { writeFavicons } from './writeFavicons.js';
import { writeSanityRuntime } from './writeSanityRuntime.js';
/**
 * Builds static files
 *
 * @internal
 */ export async function buildStaticFiles(options) {
    const { basePath, cwd, entry, importMap, isApp, minify = true, outputDir, reactCompiler, sourceMap = false, vite: extendViteConfig } = options;
    buildDebug('Writing Sanity runtime files');
    await writeSanityRuntime({
        basePath,
        cwd,
        entry,
        isApp,
        reactStrictMode: false,
        watch: false
    });
    buildDebug('Resolving vite config');
    const mode = 'production';
    let viteConfig = await getViteConfig({
        basePath,
        cwd,
        importMap,
        isApp,
        minify,
        mode,
        outputDir,
        reactCompiler,
        sourceMap
    });
    if (extendViteConfig) {
        viteConfig = await extendViteConfigWithUserConfig({
            command: 'build',
            mode
        }, viteConfig, extendViteConfig);
        viteConfig = await finalizeViteConfig(viteConfig);
    }
    const fromPath = path.join(cwd, 'static');
    // Copy files placed in /static to the built /static
    buildDebug(`Copying static files from ${fromPath} to output dir`);
    const staticPath = path.join(outputDir, 'static');
    await copyDir(fromPath, staticPath);
    // Write favicons, not overwriting ones that already exist, to static folder
    buildDebug('Writing favicons to output dir');
    const faviconBasePath = `${basePath.replace(/\/+$/, '')}/static`;
    await writeFavicons(faviconBasePath, staticPath);
    buildDebug('Bundling using vite');
    const bundle = await build(viteConfig);
    buildDebug('Bundling complete');
    // For typescript only - this shouldn't ever be the case given we're not watching
    if (Array.isArray(bundle) || !('output' in bundle)) {
        return {
            chunks: []
        };
    }
    const stats = [];
    for (const chunk of bundle.output){
        if (chunk.type !== 'chunk') {
            continue;
        }
        stats.push({
            modules: Object.entries(chunk.modules).map(([rawFilePath, chunkModule])=>{
                const filePath = rawFilePath.startsWith('\u0000') ? rawFilePath.slice('\u0000'.length) : rawFilePath;
                return {
                    name: path.isAbsolute(filePath) ? path.relative(cwd, filePath) : filePath,
                    originalLength: chunkModule.originalLength,
                    renderedLength: chunkModule.renderedLength
                };
            }),
            name: chunk.name
        });
    }
    return {
        chunks: stats
    };
}
