// @sanity/cli/dist/actions/build/buildVendorDependencies.js
import path from 'node:path';
import semver from 'semver';
import { build } from 'vite';
import { getLocalPackageDir, getLocalPackageVersion } from '../../util/getLocalPackageVersion.js';
import { createExternalFromImportMap } from './createExternalFromImportMap.js';
// Directory where vendor packages will be stored
const VENDOR_DIR = 'vendor';
// Define the vendor packages and their corresponding versions and entry points
const VENDOR_IMPORTS = {
    react: {
        '^19.2.0': {
            '.': './cjs/react.production.js',
            './compiler-runtime': './cjs/react-compiler-runtime.production.js',
            './jsx-dev-runtime': './cjs/react-jsx-dev-runtime.production.js',
            './jsx-runtime': './cjs/react-jsx-runtime.production.js',
            './package.json': './package.json'
        }
    },
    'react-dom': {
        '^19.2.0': {
            '.': './cjs/react-dom.production.js',
            './client': './cjs/react-dom-client.production.js',
            './package.json': './package.json',
            './server': './cjs/react-dom-server-legacy.browser.production.js',
            './server.browser': './cjs/react-dom-server-legacy.browser.production.js',
            './static': './cjs/react-dom-server.browser.production.js',
            './static.browser': './cjs/react-dom-server.browser.production.js'
        }
    }
};
const STYLED_COMPONENTS_IMPORTS = {
    'styled-components': {
        '^6.1.0': {
            '.': './dist/styled-components.browser.esm.js',
            './package.json': './package.json'
        }
    }
};
/**
 * Builds the ESM browser compatible versions of the vendor packages
 * specified in VENDOR_IMPORTS. Returns the `imports` object of an import map.
 */ export async function buildVendorDependencies({ basePath, cwd, isApp, outputDir }) {
    const entry = {};
    const imports = {};
    // If we're building an app, we don't need to build the styled-components package
    const vendorImports = isApp ? VENDOR_IMPORTS : {
        ...VENDOR_IMPORTS,
        ...STYLED_COMPONENTS_IMPORTS
    };
    // Iterate over each package and its version ranges in VENDOR_IMPORTS
    for (const [packageName, ranges] of Object.entries(vendorImports)){
        const version = await getLocalPackageVersion(packageName, cwd);
        if (!version) {
            throw new Error(`Could not get version for '${packageName}'`);
        }
        // Sort version ranges in descending order
        const sortedRanges = Object.keys(ranges).toSorted((range1, range2)=>{
            const min1 = semver.minVersion(range1);
            const min2 = semver.minVersion(range2);
            if (!min1) throw new Error(`Could not parse range '${range1}'`);
            if (!min2) throw new Error(`Could not parse range '${range2}'`);
            // sort them in reverse so we can rely on array `.find` below
            return semver.rcompare(min1.version, min2.version);
        });
        // Find the first version range that satisfies the package version
        const matchedRange = sortedRanges.find((range)=>semver.satisfies(version, range));
        if (!matchedRange) {
            const min = semver.minVersion(sortedRanges.at(-1));
            if (!min) {
                throw new Error(`Could not find a minimum version for package '${packageName}'`);
            }
            if (semver.gt(min.version, version)) {
                throw new Error(`Package '${packageName}' requires at least ${min.version}.`);
            }
            throw new Error(`Version '${version}' of package '${packageName}' is not supported yet.`);
        }
        const subpaths = ranges[matchedRange];
        // Resolve the actual package directory using Node module resolution,
        // so that hoisted packages in monorepos/workspaces are found correctly
        const packageDir = getLocalPackageDir(packageName, cwd);
        // Iterate over each subpath and its corresponding entry point
        for (const [subpath, relativeEntryPoint] of Object.entries(subpaths)){
            const specifier = path.posix.join(packageName, subpath);
            const chunkName = path.posix.join(packageName, path.relative(packageName, specifier) || 'index');
            entry[chunkName] = path.join(packageDir, relativeEntryPoint);
            imports[specifier] = path.posix.join('/', basePath, VENDOR_DIR, `${chunkName}.mjs`);
        }
    }
    // Use Vite to build the packages into the output directory
    let buildResult = await build({
        appType: 'custom',
        build: {
            commonjsOptions: {
                strictRequires: 'auto'
            },
            emptyOutDir: false,
            lib: {
                entry,
                formats: [
                    'es'
                ]
            },
            minify: true,
            outDir: path.join(outputDir, VENDOR_DIR),
            rollupOptions: {
                external: createExternalFromImportMap({
                    imports
                }),
                output: {
                    chunkFileNames: '[name]-[hash].mjs',
                    entryFileNames: '[name]-[hash].mjs',
                    exports: 'named',
                    format: 'es'
                },
                treeshake: {
                    preset: 'recommended'
                }
            }
        },
        // Define a custom cache directory so that sanity's vite cache
        // does not conflict with any potential local vite projects
        cacheDir: 'node_modules/.sanity/vite-vendor',
        configFile: false,
        define: {
            'process.env.NODE_ENV': JSON.stringify('production')
        },
        logLevel: 'silent',
        mode: 'production',
        root: cwd
    });
    buildResult = Array.isArray(buildResult) ? buildResult : [
        buildResult
    ];
    // Create a map of the original import specifiers to their hashed filenames
    const hashedImports = {};
    const output = buildResult.flatMap((i)=>i.output);
    for (const chunk of output){
        if (chunk.type === 'asset') continue;
        for (const [specifier, originalPath] of Object.entries(imports)){
            if (originalPath.endsWith(`${chunk.name}.mjs`)) {
                hashedImports[specifier] = path.posix.join('/', basePath, VENDOR_DIR, chunk.fileName);
            }
        }
    }
    return hashedImports;
}
