// @sanity/cli/dist/actions/build/getAutoUpdatesImportMap.js
const MODULES_HOST = process.env.SANITY_MODULES_HOST || (process.env.SANITY_INTERNAL_ENV === 'staging' ? 'https://sanity-cdn.work' : 'https://sanity-cdn.com');
function currentUnixTime() {
    return Math.floor(Date.now() / 1000);
}
/**
 * @internal
 */ export function getAutoUpdatesImportMap(packages, options = {}) {
    return Object.fromEntries(packages.flatMap((pkg)=>getAppAutoUpdateImportMapForPackage(pkg, options)));
}
/**
 * @internal
 */ function getAppAutoUpdateImportMapForPackage(pkg, options = {}) {
    const moduleUrl = getModuleUrl(pkg, options);
    return [
        [
            pkg.name,
            moduleUrl
        ],
        [
            `${pkg.name}/`,
            `${moduleUrl}/`
        ]
    ];
}
/**
 * @internal
 */ export function getModuleUrl(pkg, options = {}) {
    const { timestamp = currentUnixTime() } = options;
    return options.appId ? getByAppModuleUrl(pkg, {
        appId: options.appId,
        baseUrl: options.baseUrl,
        timestamp
    }) : getLegacyModuleUrl(pkg, {
        baseUrl: options.baseUrl,
        timestamp
    });
}
function getLegacyModuleUrl(pkg, options) {
    const encodedMinVer = encodeURIComponent(`^${pkg.version}`);
    return `${options.baseUrl || MODULES_HOST}/v1/modules/${rewriteScopedPackage(pkg.name)}/default/${encodedMinVer}/t${options.timestamp}`;
}
function getByAppModuleUrl(pkg, options) {
    const encodedMinVer = encodeURIComponent(`^${pkg.version}`);
    return `${options.baseUrl || MODULES_HOST}/v1/modules/by-app/${options.appId}/t${options.timestamp}/${encodedMinVer}/${rewriteScopedPackage(pkg.name)}`;
}
/**
 * replaces '/' with '__' similar to how eg `@types/scope__pkg` are rewritten
 * scoped packages are stored this way both in the manifest and in the cloud storage bucket
 */ function rewriteScopedPackage(pkgName) {
    if (!pkgName.includes('@')) {
        return pkgName;
    }
    const [scope, ...pkg] = pkgName.split('/');
    return `${scope}__${pkg.join('')}`;
}
