// @sanity/cli/dist/commands/dataset/embeddings/status.js
import { Args } from '@oclif/core';
import { SanityCommand, subdebug } from '@sanity/cli-core';
import { resolveDataset } from '../../../actions/dataset/resolveDataset.js';
import { promptForProject } from '../../../prompts/promptForProject.js';
import { getEmbeddingsSettings } from '../../../services/embeddings.js';
import { getProjectIdFlag } from '../../../util/sharedFlags.js';
const debug = subdebug('dataset:embeddings:status');
export class DatasetEmbeddingsStatusCommand extends SanityCommand {
    static args = {
        dataset: Args.string({
            description: 'The name of the dataset to check embeddings status for',
            required: false
        })
    };
    static description = 'Show embeddings settings and status for a dataset';
    static examples = [
        {
            command: '<%= config.bin %> <%= command.id %> production',
            description: 'Show embeddings status for the production dataset'
        }
    ];
    static flags = {
        ...getProjectIdFlag({
            description: 'Project ID to check embeddings status for',
            semantics: 'override'
        })
    };
    async run() {
        const { args } = await this.parse(DatasetEmbeddingsStatusCommand);
        let { dataset } = args;
        const projectId = await this.getProjectId({
            fallback: ()=>promptForProject({
                    requiredPermissions: [
                        {
                            grant: 'read',
                            permission: 'sanity.project.datasets'
                        }
                    ]
                })
        });
        try {
            ;
            ({ dataset } = await resolveDataset({
                dataset,
                projectId
            }));
        } catch (error) {
            const message = error instanceof Error ? error.message : String(error);
            debug(`Failed to resolve dataset: ${message}`, error);
            this.error(message, {
                exit: 1
            });
        }
        try {
            const settings = await getEmbeddingsSettings(projectId, dataset);
            this.log(`Dataset:    ${dataset}`);
            this.log(`Embeddings: ${settings.enabled ? 'enabled' : 'disabled'}`);
            if (settings.projection) {
                this.log(`Projection: ${settings.projection}`);
            }
            this.log(`Status:     ${settings.status}`);
        } catch (error) {
            const message = error instanceof Error ? error.message : String(error);
            debug(`Failed to get embeddings settings: ${message}`, error);
            this.error(`Failed to get embeddings settings: ${message}`, {
                exit: 1
            });
        }
    }
}
