// @sanity/cli/dist/util/packageManager/installPackages.js
import { spinner } from '@sanity/cli-core/ux';
import { execa } from 'execa';
import { getPartialEnvWithNpmPath } from './packageManagerChoice.js';
const PACKAGE_MANAGER_COMMANDS = {
    add: {
        bun: (packages)=>[
                'add',
                ...packages
            ],
        npm: (packages)=>[
                'install',
                '--save',
                ...packages
            ],
        pnpm: (packages)=>[
                'add',
                '--save-prod',
                ...packages
            ],
        yarn: (packages)=>[
                'add',
                ...packages
            ]
    },
    install: {
        bun: [
            'install'
        ],
        npm: [
            'install'
        ],
        pnpm: [
            'install'
        ],
        yarn: [
            'install'
        ]
    }
};
async function executePackageManagerCommand(packageManager, args, execOptions, output, errorMessage) {
    const progress = spinner(`Running ${packageManager} ${args.join(' ')}\n`).start();
    const result = await execa(packageManager, args, execOptions);
    if (result?.exitCode || result?.failed) {
        progress.fail();
        output.log(String(result.stdout));
        output.error(errorMessage, {
            exit: 1
        });
    } else {
        progress.succeed();
    }
}
export async function installDeclaredPackages(cwd, packageManager, context) {
    const { output } = context;
    const execOptions = {
        cwd,
        encoding: 'utf8',
        env: getPartialEnvWithNpmPath(cwd),
        stdio: 'pipe'
    };
    if (packageManager === 'manual') {
        const npmCommand = PACKAGE_MANAGER_COMMANDS.install.npm;
        output.log(`Manual installation selected — run 'npm ${npmCommand.join(' ')}' or equivalent`);
    } else {
        const args = PACKAGE_MANAGER_COMMANDS.install[packageManager];
        await executePackageManagerCommand(packageManager, args, execOptions, output, 'Dependency installation failed');
    }
}
export async function installNewPackages(options, context) {
    const { packageManager, packages } = options;
    const { output, workDir } = context;
    const execOptions = {
        cwd: workDir,
        encoding: 'utf8',
        env: getPartialEnvWithNpmPath(workDir),
        stdio: 'pipe'
    };
    if (packageManager === 'manual') {
        const npmCommand = PACKAGE_MANAGER_COMMANDS.add.npm(packages);
        output.log(`Manual installation selected - run 'npm ${npmCommand.join(' ')}' or equivalent`);
    } else {
        const args = PACKAGE_MANAGER_COMMANDS.add[packageManager](packages);
        await executePackageManagerCommand(packageManager, args, execOptions, output, 'Package installation failed');
    }
}
