// @sanity/cli/dist/commands/exec.js
import { Args, Flags } from '@oclif/core';
import { SanityCommand } from '@sanity/cli-core';
import { execScript } from '../actions/exec/execScript.js';
export class ExecCommand extends SanityCommand {
    static args = {
        script: Args.file({
            description: 'Path to the script to execute',
            exists: true,
            required: true
        })
    };
    static description = 'Executes a script within the Sanity Studio context';
    static examples = [
        {
            command: '<%= config.bin %> <%= command.id %> some/script.js',
            description: 'Run the script at some/script.js in Sanity context'
        },
        {
            command: '<%= config.bin %> <%= command.id %> migrations/fullname.ts --with-user-token',
            description: "Run the script at migrations/fullname.ts and configure `getCliClient()` from `sanity/cli` to include the current user's token"
        },
        {
            command: '<%= config.bin %> <%= command.id %> scripts/browserScript.js --mock-browser-env',
            description: 'Run the script at scripts/browserScript.js in a mock browser environment'
        },
        {
            command: '<%= config.bin %> <%= command.id %> --mock-browser-env myscript.js -- --dry-run positional-argument',
            description: "Pass arbitrary arguments to scripts by separating them with a `--`. Arguments are available in `process.argv` as they would in regular node scripts (eg the following command would yield a `process.argv` of: `['/path/to/node', '/path/to/myscript.js', '--dry-run', 'positional-argument']`)"
        }
    ];
    static flags = {
        'mock-browser-env': Flags.boolean({
            default: false,
            description: 'Mocks a browser-like environment using jsdom'
        }),
        'with-user-token': Flags.boolean({
            default: false,
            description: 'Prime access token from CLI config into getCliClient()'
        })
    };
    static strict = false;
    async run() {
        const { args, argv, flags } = await this.parse(ExecCommand);
        const { directory: workDir } = await this.getProjectRoot();
        await execScript({
            extraArguments: argv.slice(1),
            flags,
            scriptPath: args.script,
            workDir
        });
    }
}
