// @sanity/cli/dist/commands/login.js
import { Flags } from '@oclif/core';
import { SanityCommand } from '@sanity/cli-core';
import { login } from '../actions/auth/login/login.js';
export class LoginCommand extends SanityCommand {
    static description = 'Authenticates the CLI for access to Sanity projects';
    static examples = [
        {
            command: '<%= config.bin %> <%= command.id %>',
            description: 'Log in using default settings'
        },
        {
            command: '<%= config.bin %> <%= command.id %> --sso my-organization',
            description: 'Log in using Single Sign-On with the "my-organization" slug'
        },
        {
            command: '<%= config.bin %> <%= command.id %> --provider github --no-open',
            description: 'Login with GitHub provider, but do not open a browser window automatically'
        }
    ];
    static flags = {
        experimental: Flags.boolean({
            default: false,
            hidden: true
        }),
        open: Flags.boolean({
            allowNo: true,
            default: true,
            description: 'Open a browser window to log in (`--no-open` only prints URL)'
        }),
        provider: Flags.string({
            description: 'Log in using the given provider',
            exclusive: [
                'sso'
            ],
            helpValue: '<providerId>'
        }),
        sso: Flags.string({
            description: 'Log in using Single Sign-On, using the given organization slug',
            exclusive: [
                'provider'
            ],
            helpValue: '<slug>'
        })
    };
    async run() {
        const { flags } = await this.parse(LoginCommand);
        try {
            await login({
                ...flags,
                output: this.output,
                telemetry: this.telemetry
            });
            this.log('Login successful');
        } catch (error) {
            const message = error instanceof Error ? error.message : String(error);
            this.error(`Login failed: ${message}`, {
                exit: 1
            });
        }
    }
}
