// @sanity/vision/lib/_chunks-es/SanityVision.js
import { jsx, jsxs, Fragment as Fragment$1 } from "react/jsx-runtime";
import { c } from "react/compiler-runtime";
import { useTranslation, useKeyValueStore, useDateTimeFormat, ContextMenuButton, Translate, usePerspective, useDataset, useScheduledDraftsEnabled, useActiveReleases, useWorkspace, sortReleases, useClient, isCardinalityOneRelease, getReleaseIdFromReleaseDocumentId } from "sanity";
import { Spinner, rem, useTheme, Flex, Card, Label, Box, Text, Tooltip, Stack, useToast, TextInput, MenuButton, Menu, MenuItem, Code, Button, Hotkeys, useClickOutsideEvent, Popover, Inline, Badge, Grid, Select, Container, Heading } from "@sanity/ui";
import { useState, useEffect, forwardRef, useRef, useImperativeHandle, Fragment, useEffectEvent, Component } from "react";
import { SplitPane } from "@rexxars/react-split-pane";
import { ErrorOutlineIcon, TrashIcon, AddIcon, SearchIcon, StopIcon, PlayIcon, HelpCircleIcon, CopyIcon, LinkIcon, DocumentSheetIcon, ChevronLeftIcon, ChevronRightIcon } from "@sanity/icons";
import { isHotkey } from "is-hotkey-esm";
import CodeMirror, { EditorSelection } from "@uiw/react-codemirror";
import { closeBrackets } from "@codemirror/autocomplete";
import { history, defaultKeymap, historyKeymap } from "@codemirror/commands";
import { javascriptLanguage } from "@codemirror/lang-javascript";
import { indentOnInput, bracketMatching, syntaxHighlighting, defaultHighlightStyle, HighlightStyle } from "@codemirror/language";
import { highlightSelectionMatches } from "@codemirror/search";
import { lineNumbers, highlightActiveLine, highlightActiveLineGutter, highlightSpecialChars, drawSelection, keymap, EditorView } from "@codemirror/view";
import { tags } from "@lezer/highlight";
import { hues } from "@sanity/color";
import { styled, css } from "styled-components";
import { visionLocaleNamespace } from "./index.js";
import debounce from "lodash-es/debounce.js";
import JSON5 from "json5";
import isEqual from "lodash-es/isEqual.js";
import { uuid } from "@sanity/uuid";
import { startWith, map } from "rxjs/operators";
import { json2csv } from "json-2-csv";
import { JsonInspector } from "@rexxars/react-json-inspector";
import LRU from "quick-lru";
import { IntentLink } from "sanity/router";
import { useObservable } from "react-rx";
import { of, map as map$1, catchError } from "rxjs";
const API_VERSIONS = ["v1", "vX", "v2021-03-25", "v2021-10-21", "v2022-03-07", "v2025-02-19", `v${(/* @__PURE__ */ new Date()).toISOString().split("T")[0]}`], [DEFAULT_API_VERSION] = API_VERSIONS.slice(-1);
function DelayedSpinner(props) {
  const $ = c(5), [show, setShow] = useState(!1);
  let t0, t1;
  $[0] !== props.delay ? (t0 = () => {
    const timer = setTimeout(() => setShow(!0), props.delay || 500);
    return () => clearTimeout(timer);
  }, t1 = [props.delay], $[0] = props.delay, $[1] = t0, $[2] = t1) : (t0 = $[1], t1 = $[2]), useEffect(t0, t1);
  let t2;
  return $[3] !== show ? (t2 = show ? /* @__PURE__ */ jsx(Spinner, { muted: !0, size: 4 }) : null, $[3] = show, $[4] = t2) : t2 = $[4], t2;
}
const codemirrorExtensions = [[javascriptLanguage], lineNumbers(), highlightActiveLine(), highlightActiveLineGutter(), highlightSelectionMatches(), highlightSpecialChars(), indentOnInput(), bracketMatching(), closeBrackets(), history(), drawSelection(), syntaxHighlighting(defaultHighlightStyle, {
  fallback: !0
}), keymap.of([
  // Override the default keymap for Mod-Enter to not insert a new line, we have a custom event handler for executing queries
  {
    key: "Mod-Enter",
    run: () => !0
  },
  // Add the default keymap and history keymap
  defaultKeymap,
  historyKeymap
].flat().filter(Boolean))];
function useCodemirrorTheme(theme) {
  const $ = c(7);
  let t0;
  $[0] !== theme ? (t0 = createTheme(theme), $[0] = theme, $[1] = t0) : t0 = $[1];
  const cmTheme = t0;
  let t1;
  $[2] !== theme ? (t1 = syntaxHighlighting(createHighlight(theme)), $[2] = theme, $[3] = t1) : t1 = $[3];
  const cmHighlight = t1;
  let t2;
  return $[4] !== cmHighlight || $[5] !== cmTheme ? (t2 = [cmTheme, cmHighlight], $[4] = cmHighlight, $[5] = cmTheme, $[6] = t2) : t2 = $[6], t2;
}
function createTheme(theme) {
  const {
    color,
    fonts
  } = theme.sanity, card = color.card.enabled, cursor = hues.blue[color.dark ? 400 : 500].hex, selection = hues.gray[theme.sanity.color.dark ? 900 : 100].hex;
  return EditorView.theme({
    "&": {
      color: card.fg,
      backgroundColor: card.bg
    },
    ".cm-content": {
      caretColor: cursor
    },
    ".cm-editor": {
      fontFamily: fonts.code.family,
      fontSize: rem(fonts.code.sizes[1].fontSize),
      lineHeight: "inherit"
    },
    ".cm-cursor, .cm-dropCursor": {
      borderLeftColor: cursor
    },
    "&.cm-focused .cm-selectionBackground, .cm-selectionBackground, .cm-content ::selection": {
      backgroundColor: selection
    },
    ".cm-panels": {
      backgroundColor: card.bg,
      color: card.fg
    },
    ".cm-panels.cm-panels-top": {
      borderBottom: `2px solid ${card.border}`
    },
    ".cm-panels.cm-panels-bottom": {
      borderTop: `2px solid ${card.border}`
    }
  }, {
    dark: color.dark
  });
}
function createHighlight(theme) {
  const c2 = theme.sanity.color.base, s = theme.sanity.color.syntax;
  return HighlightStyle.define([{
    tag: tags.keyword,
    color: s.keyword
  }, {
    tag: [tags.propertyName, tags.name, tags.deleted, tags.character, tags.macroName],
    color: s.property
  }, {
    tag: [tags.function(tags.variableName), tags.labelName],
    color: s.function
  }, {
    tag: [tags.color, tags.constant(tags.name), tags.standard(tags.name)],
    color: s.variable
  }, {
    tag: [tags.definition(tags.name), tags.separator],
    color: s.constant
  }, {
    tag: [tags.typeName, tags.className, tags.number, tags.changed, tags.annotation, tags.modifier, tags.self, tags.namespace],
    color: s.number
  }, {
    tag: [tags.operator, tags.operatorKeyword, tags.url, tags.escape, tags.regexp, tags.link, tags.special(tags.string)],
    color: s.operator
  }, {
    tag: [tags.meta, tags.comment],
    color: s.comment
  }, {
    tag: tags.strong,
    fontWeight: "bold"
  }, {
    tag: tags.emphasis,
    fontStyle: "italic"
  }, {
    tag: tags.strikethrough,
    textDecoration: "line-through"
  }, {
    tag: tags.heading,
    fontWeight: "bold",
    color: s.property
  }, {
    tag: [tags.atom, tags.bool, tags.special(tags.variableName)],
    color: s.boolean
  }, {
    tag: [tags.processingInstruction, tags.string, tags.inserted],
    color: s.string
  }, {
    tag: tags.invalid,
    color: c2.fg
  }]);
}
const EditorRoot = styled.div.withConfig({
  displayName: "EditorRoot",
  componentId: "sc-1e01cjn-0"
})`width:100%;box-sizing:border-box;height:100%;overflow:hidden;overflow:clip;position:relative;display:flex;& .cm-theme{width:100%;}& .cm-editor{height:100%;font-size:16px;line-height:21px;}& .cm-line{padding-left:${({
  theme
}) => rem(theme.sanity.space[3])};}& .cm-content{border-right-width:${({
  theme
}) => rem(theme.sanity.space[4])} !important;padding-top:${({
  theme
}) => rem(theme.sanity.space[5])};}`, VisionCodeMirror = forwardRef((props, ref) => {
  const $ = c(7), [initialValue] = useState(props.initialValue), sanityTheme = useTheme(), theme = useCodemirrorTheme(sanityTheme), codeMirrorRef = useRef(null);
  let t0;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t0 = (newContent) => {
    const editorView = codeMirrorRef.current?.view;
    if (!editorView)
      return;
    const currentDoc = editorView.state.doc.toString();
    newContent !== currentDoc && editorView.dispatch({
      changes: {
        from: 0,
        to: currentDoc.length,
        insert: newContent
      },
      selection: EditorSelection.cursor(newContent.length)
    });
  }, $[0] = t0) : t0 = $[0];
  const resetEditorContent = t0;
  let t1, t2;
  $[1] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t1 = () => ({
    resetEditorContent
  }), t2 = [resetEditorContent], $[1] = t1, $[2] = t2) : (t1 = $[1], t2 = $[2]), useImperativeHandle(ref, t1, t2);
  let t3;
  return $[3] !== initialValue || $[4] !== props.onChange || $[5] !== theme ? (t3 = /* @__PURE__ */ jsx(EditorRoot, { children: /* @__PURE__ */ jsx(CodeMirror, { ref: codeMirrorRef, basicSetup: !1, theme, extensions: codemirrorExtensions, value: initialValue, onChange: props.onChange }) }), $[3] = initialValue, $[4] = props.onChange, $[5] = theme, $[6] = t3) : t3 = $[6], t3;
});
VisionCodeMirror.displayName = "VisionCodeMirror";
const SUPPORTED_PERSPECTIVES = ["pinnedRelease", "scheduledDrafts", "raw", "published", "drafts"], VIRTUAL_PERSPECTIVES = ["pinnedRelease", "scheduledDrafts"];
function isSupportedPerspective(p) {
  return SUPPORTED_PERSPECTIVES.includes(p);
}
function isVirtualPerspective(maybeVirtualPerspective) {
  return typeof maybeVirtualPerspective == "string" && VIRTUAL_PERSPECTIVES.includes(maybeVirtualPerspective);
}
function hasPinnedPerspective({
  selectedPerspectiveName
}) {
  return typeof selectedPerspectiveName < "u";
}
function getActivePerspective({
  visionPerspective,
  perspectiveStack,
  scheduledDraftsStack
}) {
  return visionPerspective === "pinnedRelease" ? perspectiveStack : visionPerspective === "scheduledDrafts" ? scheduledDraftsStack : visionPerspective;
}
function encodeQueryString(query, params = {}, options = {}) {
  const searchParams = new URLSearchParams();
  searchParams.set("query", query);
  for (const [key, value] of Object.entries(params))
    searchParams.set(`$${key}`, JSON.stringify(value));
  for (const [key, value] of Object.entries(options))
    value && searchParams.set(key, `${value}`);
  return `?${searchParams}`;
}
function isPlainObject(obj) {
  return !!obj && typeof obj == "object" && Object.prototype.toString.call(obj) === "[object Object]";
}
const hasLocalStorage = supportsLocalStorage(), keyPrefix = "sanityVision:";
function clearLocalStorage() {
  if (hasLocalStorage)
    for (let i = 0; i < localStorage.length; i++) {
      const key = localStorage.key(i);
      key?.startsWith(keyPrefix) && localStorage.removeItem(key);
    }
}
function getLocalStorage(namespace) {
  const storageKey = `${keyPrefix}${namespace}`;
  let loadedState = null;
  return {
    get,
    set,
    merge
  };
  function get(key, defaultVal) {
    const state = ensureState();
    return typeof state[key] > "u" ? defaultVal : state[key];
  }
  function set(key, value) {
    const state = ensureState();
    return state[key] = value, localStorage.setItem(storageKey, JSON.stringify(loadedState)), value;
  }
  function merge(props) {
    const state = {
      ...ensureState(),
      ...props
    };
    return localStorage.setItem(storageKey, JSON.stringify(state)), state;
  }
  function ensureState() {
    return loadedState === null && (loadedState = loadState()), loadedState;
  }
  function loadState() {
    if (!hasLocalStorage)
      return {};
    try {
      const stored = JSON.parse(localStorage.getItem(storageKey) || "{}");
      return isPlainObject(stored) ? stored : {};
    } catch {
      return {};
    }
  }
}
function supportsLocalStorage() {
  const mod = "lsCheck";
  try {
    return localStorage.setItem(mod, mod), localStorage.removeItem(mod), !0;
  } catch {
    return !1;
  }
}
function parseApiQueryString(qs) {
  const params = {}, options = {};
  for (const [key, value] of qs.entries()) {
    if (key[0] === "$") {
      params[key.slice(1)] = JSON.parse(value);
      continue;
    }
    if (key === "perspective") {
      options[key] = value;
      continue;
    }
  }
  return {
    query: qs.get("query") || "",
    params,
    options
  };
}
function prefixApiVersion(version) {
  return version[0] !== "v" && version !== "other" ? `v${version}` : version;
}
function validateApiVersion(apiVersion) {
  const parseableApiVersion = apiVersion.replace(/^v/, "").trim().toUpperCase();
  return parseableApiVersion.length > 0 && (parseableApiVersion === "X" || parseableApiVersion === "1" || /^\d{4}-\d{2}-\d{2}$/.test(parseableApiVersion) && !isNaN(Date.parse(parseableApiVersion)));
}
function tryParseParams(val, t) {
  try {
    const parsed = val ? JSON5.parse(val) : {};
    return typeof parsed == "object" && parsed && !Array.isArray(parsed) ? parsed : {};
  } catch (err) {
    return err.message = `${t("params.error.params-invalid-json")}:

${err.message.replace("JSON5:", "")}`, err;
  }
}
const Root = styled(Flex).withConfig({
  displayName: "Root",
  componentId: "sc-r315za-0"
})`.sidebarPanes .Pane{overflow-y:auto;overflow-x:hidden;}& .Resizer{background:var(--card-border-color);opacity:1;z-index:1;box-sizing:border-box;background-clip:padding-box;border:solid transparent;}& .Resizer:hover{border-color:var(--card-shadow-ambient-color);}& .Resizer.horizontal{height:11px;margin:-5px 0;border-width:5px 0;cursor:row-resize;width:100%;z-index:4;}& .Resizer.vertical{width:11px;margin:0 -5px;border-width:0 5px;cursor:col-resize;z-index:2;}.Resizer.disabled{cursor:not-allowed;}.Resizer.disabled:hover{border-color:transparent;}`;
Root.displayName = "Root";
const Header = styled(Card).withConfig({
  displayName: "Header",
  componentId: "sc-r315za-1"
})`border-bottom:1px solid var(--card-border-color);`, StyledLabel = styled(Label).withConfig({
  displayName: "StyledLabel",
  componentId: "sc-r315za-2"
})`flex:1;`, SplitpaneContainer = styled(Box).withConfig({
  displayName: "SplitpaneContainer",
  componentId: "sc-r315za-3"
})`position:relative;`, QueryCopyLink = styled.a.withConfig({
  displayName: "QueryCopyLink",
  componentId: "sc-r315za-4"
})`cursor:pointer;margin-right:auto;`, InputBackgroundContainer = styled(Box).withConfig({
  displayName: "InputBackgroundContainer",
  componentId: "sc-r315za-5"
})`position:absolute;top:1rem;left:0;padding:0;margin:0;z-index:10;right:0;${StyledLabel}{user-select:none;}`, InputBackgroundContainerLeft = styled(InputBackgroundContainer).withConfig({
  displayName: "InputBackgroundContainerLeft",
  componentId: "sc-r315za-6"
})`left:33px;`, InputContainer = styled(Card).withConfig({
  displayName: "InputContainer",
  componentId: "sc-r315za-7"
})`width:100%;height:100%;position:relative;flex-direction:column;`, ResultOuterContainer = styled(Flex).withConfig({
  displayName: "ResultOuterContainer",
  componentId: "sc-r315za-8"
})`height:100%;`, ResultInnerContainer = styled(Box).withConfig({
  displayName: "ResultInnerContainer",
  componentId: "sc-r315za-9"
})`position:relative;`, ResultContainer = styled(Card).withConfig({
  displayName: "ResultContainer",
  componentId: "sc-r315za-10"
})`height:100%;width:100%;position:absolute;max-width:100%;${({
  $isInvalid
}) => $isInvalid && css`
      &:after {
        background-color: var(--card-bg-color);
        content: '';
        position: absolute;
        top: 0;
        bottom: 0;
        left: 0;
        width: 100%;
      }
    `}`, Result = styled(Box).withConfig({
  displayName: "Result",
  componentId: "sc-r315za-11"
})`position:relative;width:100%;height:100%;z-index:20;`, ResultFooter = styled(Flex).withConfig({
  displayName: "ResultFooter",
  componentId: "sc-r315za-12"
})`border-top:1px solid var(--card-border-color);`, TimingsCard = styled(Card).withConfig({
  displayName: "TimingsCard",
  componentId: "sc-r315za-13"
})`position:relative;`, TimingsTextContainer = styled(Flex).withConfig({
  displayName: "TimingsTextContainer",
  componentId: "sc-r315za-14"
})`height:100%;min-height:${({
  theme
}) => rem(theme.sanity.space[3] * 2 + theme.sanity.fonts.text.sizes[2].lineHeight - theme.sanity.fonts.text.sizes[2].ascenderHeight - theme.sanity.fonts.text.sizes[2].descenderHeight)};`, DownloadsCard = styled(Card).withConfig({
  displayName: "DownloadsCard",
  componentId: "sc-r315za-15"
})`position:relative;`, SaveResultLabel = styled(Text).withConfig({
  displayName: "SaveResultLabel",
  componentId: "sc-r315za-16"
})`transform:initial;&:before,&:after{content:none;}> span{display:flex !important;gap:${({
  theme
}) => rem(theme.sanity.space[3])};align-items:center;}`, ControlsContainer = styled(Box).withConfig({
  displayName: "ControlsContainer",
  componentId: "sc-r315za-17"
})`border-top:1px solid var(--card-border-color);`, defaultValue$1 = `{
  
}`;
function ParamsEditor(props) {
  const $ = c(21), {
    onChange,
    paramsError,
    hasValidParams,
    editorRef
  } = props, {
    t
  } = useTranslation(visionLocaleNamespace);
  let t0;
  $[0] !== onChange ? (t0 = (newValue) => {
    onChange(newValue);
  }, $[0] = onChange, $[1] = t0) : t0 = $[1];
  const handleChangeRaw = t0;
  let t1;
  $[2] !== handleChangeRaw ? (t1 = debounce(handleChangeRaw, 333), $[2] = handleChangeRaw, $[3] = t1) : t1 = $[3];
  const handleChange = t1, t2 = hasValidParams ? "default" : "critical";
  let t3;
  $[4] !== t ? (t3 = t("params.label"), $[4] = t, $[5] = t3) : t3 = $[5];
  let t4;
  $[6] !== t3 ? (t4 = /* @__PURE__ */ jsx(StyledLabel, { muted: !0, children: t3 }), $[6] = t3, $[7] = t4) : t4 = $[7];
  let t5;
  $[8] !== paramsError ? (t5 = paramsError && /* @__PURE__ */ jsx(Tooltip, { animate: !0, placement: "top", portal: !0, content: /* @__PURE__ */ jsx(Text, { size: 1, children: paramsError }), children: /* @__PURE__ */ jsx(Box, { padding: 1, marginX: 2, children: /* @__PURE__ */ jsx(Text, { children: /* @__PURE__ */ jsx(ErrorOutlineIcon, {}) }) }) }), $[8] = paramsError, $[9] = t5) : t5 = $[9];
  let t6;
  $[10] !== t4 || $[11] !== t5 ? (t6 = /* @__PURE__ */ jsx(InputBackgroundContainerLeft, { children: /* @__PURE__ */ jsxs(Flex, { children: [
    t4,
    t5
  ] }) }), $[10] = t4, $[11] = t5, $[12] = t6) : t6 = $[12];
  const t7 = props.value || defaultValue$1;
  let t8;
  $[13] !== editorRef || $[14] !== handleChange || $[15] !== t7 ? (t8 = /* @__PURE__ */ jsx(VisionCodeMirror, { ref: editorRef, initialValue: t7, onChange: handleChange }), $[13] = editorRef, $[14] = handleChange, $[15] = t7, $[16] = t8) : t8 = $[16];
  let t9;
  return $[17] !== t2 || $[18] !== t6 || $[19] !== t8 ? (t9 = /* @__PURE__ */ jsxs(Card, { flex: 1, tone: t2, "data-testid": "params-editor", children: [
    t6,
    t8
  ] }), $[17] = t2, $[18] = t6, $[19] = t8, $[20] = t9) : t9 = $[20], t9;
}
function parseParams(value, t) {
  const parsedParams = tryParseParams(value, t), params = parsedParams instanceof Error ? {} : parsedParams, validationError = parsedParams instanceof Error ? parsedParams.message : void 0;
  return {
    parsed: params,
    raw: value,
    valid: !validationError,
    error: validationError
  };
}
const STORED_QUERIES_NAMESPACE = "studio.vision-tool.saved-queries", defaultValue = {
  queries: []
}, keyValueStoreKey = STORED_QUERIES_NAMESPACE;
function useSavedQueries() {
  const $ = c(27), keyValueStore = useKeyValueStore(), [value, setValue] = useState(defaultValue), [saving, setSaving] = useState(!1);
  let t0;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t0 = [], $[0] = t0) : t0 = $[0];
  const [deleting, setDeleting] = useState(t0), [saveQueryError, setSaveQueryError] = useState(), [deleteQueryError, setDeleteQueryError] = useState(), [error, setError] = useState();
  let t1;
  $[1] !== keyValueStore ? (t1 = keyValueStore.getKey(keyValueStoreKey), $[1] = keyValueStore, $[2] = t1) : t1 = $[2];
  const queries = t1;
  let t2;
  $[3] !== queries ? (t2 = () => {
    const sub = queries.pipe(startWith(defaultValue), map(_temp$5)).subscribe({
      next: setValue,
      error: (err) => setError(err)
    });
    return () => sub?.unsubscribe();
  }, $[3] = queries, $[4] = t2) : t2 = $[4];
  let t3;
  $[5] !== keyValueStore || $[6] !== queries ? (t3 = [queries, keyValueStore], $[5] = keyValueStore, $[6] = queries, $[7] = t3) : t3 = $[7], useEffect(t2, t3);
  let t4;
  $[8] !== keyValueStore || $[9] !== value.queries ? (t4 = async (query) => {
    setSaving(!0), setSaveQueryError(void 0);
    try {
      const newQueries = [{
        ...query,
        _key: uuid()
      }, ...value.queries];
      setValue({
        queries: newQueries
      }), await keyValueStore.setKey(keyValueStoreKey, {
        queries: newQueries
      });
    } catch (t52) {
      setSaveQueryError(t52);
    }
    setSaving(!1);
  }, $[8] = keyValueStore, $[9] = value.queries, $[10] = t4) : t4 = $[10];
  const saveQuery = t4;
  let t5;
  $[11] !== keyValueStore || $[12] !== value.queries ? (t5 = async (query_0) => {
    setSaving(!0), setSaveQueryError(void 0);
    try {
      const updatedQueries = value.queries.map((q) => q._key === query_0._key ? {
        ...q,
        ...query_0
      } : q);
      setValue({
        queries: updatedQueries
      }), await keyValueStore.setKey(keyValueStoreKey, {
        queries: updatedQueries
      });
    } catch (t62) {
      setSaveQueryError(t62);
    }
    setSaving(!1);
  }, $[11] = keyValueStore, $[12] = value.queries, $[13] = t5) : t5 = $[13];
  const updateQuery = t5;
  let t6;
  $[14] !== keyValueStore || $[15] !== value.queries ? (t6 = async (key) => {
    setDeleting((prev) => [...prev, key]), setDeleteQueryError(void 0);
    try {
      const filteredQueries = value.queries.filter((q_0) => q_0._key !== key);
      setValue({
        queries: filteredQueries
      }), await keyValueStore.setKey(keyValueStoreKey, {
        queries: filteredQueries
      });
    } catch (t72) {
      setDeleteQueryError(t72);
    }
    setDeleting((prev_0) => prev_0.filter((k) => k !== key));
  }, $[14] = keyValueStore, $[15] = value.queries, $[16] = t6) : t6 = $[16];
  const deleteQuery = t6;
  let t7;
  return $[17] !== deleteQuery || $[18] !== deleteQueryError || $[19] !== deleting || $[20] !== error || $[21] !== saveQuery || $[22] !== saveQueryError || $[23] !== saving || $[24] !== updateQuery || $[25] !== value.queries ? (t7 = {
    queries: value.queries,
    saveQuery,
    updateQuery,
    deleteQuery,
    saving,
    deleting,
    saveQueryError,
    deleteQueryError,
    error
  }, $[17] = deleteQuery, $[18] = deleteQueryError, $[19] = deleting, $[20] = error, $[21] = saveQuery, $[22] = saveQueryError, $[23] = saving, $[24] = updateQuery, $[25] = value.queries, $[26] = t7) : t7 = $[26], t7;
}
function _temp$5(data) {
  return data || defaultValue;
}
const FixedHeader = styled(Stack).withConfig({
  displayName: "FixedHeader",
  componentId: "sc-14q7beb-0"
})`position:sticky;top:0;background:${({
  theme
}) => theme.sanity.color.base.bg};z-index:1;`, ScrollContainer = styled(Box).withConfig({
  displayName: "ScrollContainer",
  componentId: "sc-14q7beb-1"
})`height:100%;overflow-y:auto;overflow-x:hidden;&::-webkit-scrollbar{width:8px;}&::-webkit-scrollbar-track{background:transparent;}&::-webkit-scrollbar-thumb{background:${({
  theme
}) => theme.sanity.color.base.border};border-radius:4px;}`;
function QueryRecall(t0) {
  const $ = c(80), {
    url,
    getStateFromUrl,
    setStateFromParsedUrl,
    currentQuery,
    currentParams,
    generateUrl
  } = t0, toast = useToast(), {
    saveQuery,
    updateQuery,
    queries,
    deleteQuery,
    saving,
    saveQueryError
  } = useSavedQueries(), {
    t
  } = useTranslation(visionLocaleNamespace);
  let t1;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t1 = {
    month: "short",
    day: "numeric",
    year: "numeric",
    hour: "numeric",
    minute: "2-digit",
    hour12: !0
  }, $[0] = t1) : t1 = $[0];
  const formatDate = useDateTimeFormat(t1), [editingKey, setEditingKey] = useState(null), [editingTitle, setEditingTitle] = useState("");
  let t2;
  $[1] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t2 = {}, $[1] = t2) : t2 = $[1];
  const [optimisticTitles, setOptimisticTitles] = useState(t2), [searchQuery, setSearchQuery] = useState(""), [selectedUrl, setSelectedUrl] = useState(url);
  let t3;
  $[2] !== currentParams || $[3] !== currentQuery || $[4] !== formatDate || $[5] !== generateUrl || $[6] !== getStateFromUrl || $[7] !== queries || $[8] !== saveQuery || $[9] !== saveQueryError || $[10] !== t || $[11] !== toast ? (t3 = async () => {
    const newUrl = generateUrl(currentQuery, currentParams);
    if (queries?.some((q) => {
      const savedQueryObj = getStateFromUrl(q.url);
      return savedQueryObj && savedQueryObj.query === currentQuery && isEqual(savedQueryObj.params, currentParams);
    })) {
      const duplicateQuery = queries?.find((q_0) => {
        const savedQueryObj_0 = getStateFromUrl(q_0.url);
        return savedQueryObj_0 && savedQueryObj_0.query === currentQuery && isEqual(savedQueryObj_0.params, currentParams);
      });
      toast.push({
        closable: !0,
        status: "warning",
        title: t("save-query.already-saved"),
        description: `${duplicateQuery?.title} - ${formatDate.format(new Date(duplicateQuery?.savedAt || ""))}`
      });
      return;
    }
    newUrl && (await saveQuery({
      url: newUrl,
      savedAt: (/* @__PURE__ */ new Date()).toISOString(),
      title: "Untitled"
    }), setSelectedUrl(newUrl)), saveQueryError ? toast.push({
      closable: !0,
      status: "error",
      title: t("save-query.error"),
      description: saveQueryError.message
    }) : toast.push({
      closable: !0,
      status: "success",
      title: t("save-query.success")
    });
  }, $[2] = currentParams, $[3] = currentQuery, $[4] = formatDate, $[5] = generateUrl, $[6] = getStateFromUrl, $[7] = queries, $[8] = saveQuery, $[9] = saveQueryError, $[10] = t, $[11] = toast, $[12] = t3) : t3 = $[12];
  const handleSave = t3;
  let t4;
  $[13] !== t || $[14] !== toast || $[15] !== updateQuery ? (t4 = async (query, newTitle) => {
    setEditingKey(null), setOptimisticTitles((prev) => ({
      ...prev,
      [query._key]: newTitle
    }));
    try {
      await updateQuery({
        ...query,
        title: newTitle
      }), setOptimisticTitles((prev_1) => {
        const next_0 = {
          ...prev_1
        };
        return delete next_0[query._key], next_0;
      });
    } catch (t52) {
      const err = t52;
      setOptimisticTitles((prev_0) => {
        const next = {
          ...prev_0
        };
        return delete next[query._key], next;
      }), toast.push({
        closable: !0,
        status: "error",
        title: t("save-query.error"),
        description: err.message
      });
    }
  }, $[13] = t, $[14] = toast, $[15] = updateQuery, $[16] = t4) : t4 = $[16];
  const handleTitleSave = t4;
  let t5;
  $[17] !== currentParams || $[18] !== currentQuery || $[19] !== formatDate || $[20] !== generateUrl || $[21] !== getStateFromUrl || $[22] !== queries || $[23] !== t || $[24] !== toast || $[25] !== updateQuery ? (t5 = async (query_0) => {
    const newUrl_0 = generateUrl(currentQuery, currentParams);
    if (queries?.some((q_1) => {
      if (q_1._key === query_0._key)
        return !1;
      const savedQueryObj_1 = getStateFromUrl(q_1.url);
      return savedQueryObj_1 && savedQueryObj_1.query === currentQuery && isEqual(savedQueryObj_1.params, currentParams);
    })) {
      const duplicateQuery_0 = queries?.find((q_2) => {
        if (q_2._key === query_0._key)
          return !1;
        const savedQueryObj_2 = getStateFromUrl(q_2.url);
        return savedQueryObj_2 && savedQueryObj_2.query === currentQuery && isEqual(savedQueryObj_2.params, currentParams);
      });
      toast.push({
        closable: !0,
        status: "warning",
        title: t("save-query.already-saved"),
        description: `${duplicateQuery_0?.title} - ${formatDate.format(new Date(duplicateQuery_0?.savedAt || ""))}`
      });
      return;
    }
    try {
      await updateQuery({
        ...query_0,
        url: newUrl_0,
        savedAt: (/* @__PURE__ */ new Date()).toISOString()
      }), setSelectedUrl(newUrl_0), toast.push({
        closable: !0,
        status: "success",
        title: t("save-query.success")
      });
    } catch (t62) {
      const err_0 = t62;
      toast.push({
        closable: !0,
        status: "error",
        title: t("save-query.error"),
        description: err_0.message
      });
    }
  }, $[17] = currentParams, $[18] = currentQuery, $[19] = formatDate, $[20] = generateUrl, $[21] = getStateFromUrl, $[22] = queries, $[23] = t, $[24] = toast, $[25] = updateQuery, $[26] = t5) : t5 = $[26];
  const handleUpdate = t5;
  let T0, T1, t6, t7, t8;
  if ($[27] !== currentParams || $[28] !== currentQuery || $[29] !== deleteQuery || $[30] !== editingKey || $[31] !== editingTitle || $[32] !== formatDate || $[33] !== getStateFromUrl || $[34] !== handleSave || $[35] !== handleTitleSave || $[36] !== handleUpdate || $[37] !== optimisticTitles || $[38] !== queries || $[39] !== saving || $[40] !== searchQuery || $[41] !== selectedUrl || $[42] !== setStateFromParsedUrl || $[43] !== t) {
    const filteredQueries = queries?.filter((q_3) => q_3?.title?.toLowerCase().includes(searchQuery.toLowerCase()));
    T1 = ScrollContainer;
    let t92;
    $[49] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t92 = {
      textTransform: "capitalize"
    }, $[49] = t92) : t92 = $[49];
    let t102;
    $[50] !== t ? (t102 = t("label.saved-queries"), $[50] = t, $[51] = t102) : t102 = $[51];
    let t11;
    $[52] !== t102 ? (t11 = /* @__PURE__ */ jsx(Text, { weight: "semibold", style: t92, size: 4, children: t102 }), $[52] = t102, $[53] = t11) : t11 = $[53];
    let t12;
    $[54] !== t ? (t12 = t("action.save-query"), $[54] = t, $[55] = t12) : t12 = $[55];
    let t13;
    $[56] !== handleSave || $[57] !== saving || $[58] !== t12 ? (t13 = /* @__PURE__ */ jsx(Button, { label: t12, icon: AddIcon, disabled: saving, onClick: handleSave, mode: "bleed" }), $[56] = handleSave, $[57] = saving, $[58] = t12, $[59] = t13) : t13 = $[59];
    let t14;
    $[60] !== t11 || $[61] !== t13 ? (t14 = /* @__PURE__ */ jsxs(Flex, { padding: 3, paddingTop: 4, paddingBottom: 0, justify: "space-between", align: "center", children: [
      t11,
      t13
    ] }), $[60] = t11, $[61] = t13, $[62] = t14) : t14 = $[62];
    let t15;
    $[63] !== t ? (t15 = t("label.search-queries"), $[63] = t, $[64] = t15) : t15 = $[64];
    let t16;
    $[65] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t16 = (event) => setSearchQuery(event.currentTarget.value), $[65] = t16) : t16 = $[65];
    let t17;
    $[66] !== searchQuery || $[67] !== t15 ? (t17 = /* @__PURE__ */ jsx(Box, { padding: 3, paddingTop: 0, children: /* @__PURE__ */ jsx(TextInput, { placeholder: t15, icon: SearchIcon, value: searchQuery, onChange: t16 }) }), $[66] = searchQuery, $[67] = t15, $[68] = t17) : t17 = $[68], $[69] !== t14 || $[70] !== t17 ? (t8 = /* @__PURE__ */ jsxs(FixedHeader, { space: 3, children: [
      t14,
      t17
    ] }), $[69] = t14, $[70] = t17, $[71] = t8) : t8 = $[71], T0 = Stack, t6 = 3, t7 = filteredQueries?.map((q_4) => {
      const queryObj = getStateFromUrl(q_4.url), isSelected = selectedUrl === q_4.url, areQueriesEqual = queryObj && currentQuery === queryObj.query && isEqual(currentParams, queryObj.params), isEdited = isSelected && !areQueriesEqual;
      return /* @__PURE__ */ jsx(Card, { width: "fill", padding: 4, border: !0, tone: isSelected ? "positive" : "default", onClick: () => {
        setSelectedUrl(q_4.url);
        const parsedUrl = getStateFromUrl(q_4.url);
        parsedUrl && setStateFromParsedUrl(parsedUrl);
      }, style: {
        position: "relative"
      }, children: /* @__PURE__ */ jsxs(Stack, { space: 3, children: [
        /* @__PURE__ */ jsxs(Flex, { justify: "space-between", align: "center", children: [
          /* @__PURE__ */ jsxs(Flex, { align: "center", gap: 2, paddingRight: 1, children: [
            editingKey === q_4._key ? /* @__PURE__ */ jsx(TextInput, { value: editingTitle, onChange: (event_0) => setEditingTitle(event_0.currentTarget.value), onKeyDown: (event_1) => {
              event_1.key === "Enter" ? handleTitleSave(q_4, editingTitle) : event_1.key === "Escape" && setEditingKey(null);
            }, onBlur: () => handleTitleSave(q_4, editingTitle), autoFocus: !0, style: {
              maxWidth: "170px",
              height: "24px"
            } }) : /* @__PURE__ */ jsx(Text, { weight: "bold", size: 3, textOverflow: "ellipsis", style: {
              maxWidth: "170px",
              cursor: "pointer",
              padding: "4px 0"
            }, title: optimisticTitles[q_4._key] || q_4.title || q_4._key.slice(q_4._key.length - 5, q_4._key.length), onClick: () => {
              setEditingKey(q_4._key), setEditingTitle(q_4.title || q_4._key.slice(0, 5));
            }, children: optimisticTitles[q_4._key] || q_4.title || q_4._key.slice(q_4._key.length - 5, q_4._key.length) }),
            isEdited && /* @__PURE__ */ jsx(Box, { style: {
              width: "6px",
              height: "6px",
              borderRadius: "50%",
              backgroundColor: "var(--card-focus-ring-color)"
            } })
          ] }),
          /* @__PURE__ */ jsx(MenuButton, { button: /* @__PURE__ */ jsx(ContextMenuButton, {}), id: `${q_4._key}-menu`, menu: /* @__PURE__ */ jsx(Menu, { children: /* @__PURE__ */ jsx(MenuItem, { tone: "critical", padding: 3, icon: TrashIcon, text: t("action.delete"), onClick: (event_2) => {
            event_2.stopPropagation(), deleteQuery(q_4._key);
          } }) }), popover: {
            portal: !0,
            placement: "bottom-end",
            tone: "default"
          } })
        ] }),
        /* @__PURE__ */ jsx(Code, { muted: !0, children: queryObj?.query.split("{")[0] }),
        /* @__PURE__ */ jsx(Flex, { align: "center", gap: 1, children: /* @__PURE__ */ jsx(Text, { size: 1, muted: !0, children: formatDate.format(new Date(q_4.savedAt || "")) }) }),
        isEdited && /* @__PURE__ */ jsx(Button, { mode: "ghost", tone: "default", size: 1, padding: 2, style: {
          height: "24px",
          position: "absolute",
          right: "16px",
          bottom: "16px",
          fontSize: "12px"
        }, text: t("action.update"), onClick: (e) => {
          e.stopPropagation(), handleUpdate(q_4);
        } })
      ] }) }, q_4._key);
    }), $[27] = currentParams, $[28] = currentQuery, $[29] = deleteQuery, $[30] = editingKey, $[31] = editingTitle, $[32] = formatDate, $[33] = getStateFromUrl, $[34] = handleSave, $[35] = handleTitleSave, $[36] = handleUpdate, $[37] = optimisticTitles, $[38] = queries, $[39] = saving, $[40] = searchQuery, $[41] = selectedUrl, $[42] = setStateFromParsedUrl, $[43] = t, $[44] = T0, $[45] = T1, $[46] = t6, $[47] = t7, $[48] = t8;
  } else
    T0 = $[44], T1 = $[45], t6 = $[46], t7 = $[47], t8 = $[48];
  let t9;
  $[72] !== T0 || $[73] !== t6 || $[74] !== t7 ? (t9 = /* @__PURE__ */ jsx(T0, { paddingY: t6, children: t7 }), $[72] = T0, $[73] = t6, $[74] = t7, $[75] = t9) : t9 = $[75];
  let t10;
  return $[76] !== T1 || $[77] !== t8 || $[78] !== t9 ? (t10 = /* @__PURE__ */ jsxs(T1, { children: [
    t8,
    t9
  ] }), $[76] = T1, $[77] = t8, $[78] = t9, $[79] = t10) : t10 = $[79], t10;
}
function narrowBreakpoint() {
  return typeof window < "u" && window.innerWidth > 600;
}
function calculatePaneSizeOptions(height) {
  let rootHeight = height;
  return rootHeight || (rootHeight = typeof window < "u" && typeof document < "u" ? document.body.getBoundingClientRect().height - 60 : 0), {
    defaultSize: rootHeight / (narrowBreakpoint() ? 2 : 1),
    size: rootHeight > 550 ? void 0 : rootHeight * 0.4,
    allowResize: rootHeight > 550,
    minSize: Math.min(170, Math.max(170, rootHeight / 2)),
    maxSize: rootHeight > 650 ? rootHeight * 0.7 : rootHeight * 0.6
  };
}
function usePaneSize(t0) {
  const $ = c(6), {
    visionRootRef
  } = t0, [isNarrowBreakpoint, setIsNarrowBreakpoint] = useState(_temp$4), [paneSizeOptions, setPaneSizeOptions] = useState(_temp2$3);
  let t1, t2;
  $[0] !== visionRootRef ? (t1 = () => {
    if (!visionRootRef.current)
      return;
    const handleResize = (entries) => {
      setIsNarrowBreakpoint(narrowBreakpoint());
      const entry = entries?.[0];
      entry && setPaneSizeOptions(calculatePaneSizeOptions(entry.contentRect.height));
    }, resizeObserver = new ResizeObserver(handleResize);
    return resizeObserver.observe(visionRootRef.current), () => {
      resizeObserver.disconnect();
    };
  }, t2 = [visionRootRef], $[0] = visionRootRef, $[1] = t1, $[2] = t2) : (t1 = $[1], t2 = $[2]), useEffect(t1, t2);
  let t3;
  return $[3] !== isNarrowBreakpoint || $[4] !== paneSizeOptions ? (t3 = {
    paneSizeOptions,
    isNarrowBreakpoint
  }, $[3] = isNarrowBreakpoint, $[4] = paneSizeOptions, $[5] = t3) : t3 = $[5], t3;
}
function _temp2$3() {
  return calculatePaneSizeOptions(void 0);
}
function _temp$4() {
  return narrowBreakpoint();
}
function VisionGuiControls(t0) {
  const $ = c(30), {
    hasValidParams,
    listenInProgress,
    queryInProgress,
    onQueryExecution,
    onListenExecution
  } = t0, {
    t
  } = useTranslation(visionLocaleNamespace);
  let t1;
  $[0] !== t ? (t1 = t("params.error.params-invalid-json"), $[0] = t, $[1] = t1) : t1 = $[1];
  let t2;
  $[2] !== t1 ? (t2 = /* @__PURE__ */ jsx(Card, { radius: 4, children: /* @__PURE__ */ jsx(Text, { size: 1, muted: !0, children: t1 }) }), $[2] = t1, $[3] = t2) : t2 = $[3];
  let t3;
  $[4] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t3 = /* @__PURE__ */ jsx(Card, { radius: 4, children: /* @__PURE__ */ jsx(Hotkeys, { keys: ["Ctrl", "Enter"] }) }), $[4] = t3) : t3 = $[4];
  const t4 = queryInProgress ? StopIcon : PlayIcon, t5 = listenInProgress || !hasValidParams, t6 = queryInProgress ? "positive" : "primary";
  let t7;
  $[5] !== queryInProgress || $[6] !== t ? (t7 = t(queryInProgress ? "action.query-cancel" : "action.query-execute"), $[5] = queryInProgress, $[6] = t, $[7] = t7) : t7 = $[7];
  let t8;
  $[8] !== onQueryExecution || $[9] !== t4 || $[10] !== t5 || $[11] !== t6 || $[12] !== t7 ? (t8 = /* @__PURE__ */ jsx(Box, { flex: 1, children: /* @__PURE__ */ jsx(Tooltip, { content: t3, placement: "top", portal: !0, children: /* @__PURE__ */ jsx(Button, { width: "fill", onClick: onQueryExecution, type: "button", icon: t4, disabled: t5, tone: t6, text: t7 }) }) }), $[8] = onQueryExecution, $[9] = t4, $[10] = t5, $[11] = t6, $[12] = t7, $[13] = t8) : t8 = $[13];
  const t9 = listenInProgress ? StopIcon : PlayIcon;
  let t10;
  $[14] !== listenInProgress || $[15] !== t ? (t10 = t(listenInProgress ? "action.listen-cancel" : "action.listen-execute"), $[14] = listenInProgress, $[15] = t, $[16] = t10) : t10 = $[16];
  const t11 = !hasValidParams, t12 = listenInProgress ? "positive" : "default";
  let t13;
  $[17] !== onListenExecution || $[18] !== t10 || $[19] !== t11 || $[20] !== t12 || $[21] !== t9 ? (t13 = /* @__PURE__ */ jsx(Box, { flex: 1, marginLeft: 3, children: /* @__PURE__ */ jsx(Button, { width: "fill", onClick: onListenExecution, type: "button", icon: t9, text: t10, mode: "ghost", disabled: t11, tone: t12 }) }), $[17] = onListenExecution, $[18] = t10, $[19] = t11, $[20] = t12, $[21] = t9, $[22] = t13) : t13 = $[22];
  let t14;
  $[23] !== t13 || $[24] !== t8 ? (t14 = /* @__PURE__ */ jsxs(Flex, { justify: "space-evenly", children: [
    t8,
    t13
  ] }), $[23] = t13, $[24] = t8, $[25] = t14) : t14 = $[25];
  let t15;
  return $[26] !== hasValidParams || $[27] !== t14 || $[28] !== t2 ? (t15 = /* @__PURE__ */ jsx(ControlsContainer, { children: /* @__PURE__ */ jsx(Card, { padding: 3, paddingX: 3, children: /* @__PURE__ */ jsx(Tooltip, { content: t2, placement: "top", disabled: hasValidParams, portal: !0, children: t14 }) }) }), $[26] = hasValidParams, $[27] = t14, $[28] = t2, $[29] = t15) : t15 = $[29], t15;
}
const PerspectivePopoverContent = styled(Box).withConfig({
  displayName: "PerspectivePopoverContent",
  componentId: "sc-ls0jfp-0"
})`max-width:240px;`, PerspectivePopoverLink = styled.a.withConfig({
  displayName: "PerspectivePopoverLink",
  componentId: "sc-ls0jfp-1"
})`cursor:pointer;margin-right:auto;`, Dot = styled.div.withConfig({
  displayName: "Dot",
  componentId: "sc-y1ykk8-0"
})`width:4px;height:4px;border-radius:3px;box-shadow:0 0 0 1px var(--card-bg-color);background-color:${({
  $tone
}) => `var(--card-badge-${$tone}-dot-color)`};`;
function PerspectivePopover() {
  const $ = c(39), [open, setOpen] = useState(!1), buttonRef = useRef(null), popoverRef = useRef(null);
  let t0;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t0 = () => setOpen(_temp$3), $[0] = t0) : t0 = $[0];
  const handleClick = t0, {
    t
  } = useTranslation(visionLocaleNamespace);
  let t1, t2;
  $[1] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t1 = () => setOpen(!1), t2 = () => [buttonRef.current, popoverRef.current], $[1] = t1, $[2] = t2) : (t1 = $[1], t2 = $[2]), useClickOutsideEvent(t1, t2);
  let t3;
  $[3] !== t ? (t3 = t("settings.perspectives.title"), $[3] = t, $[4] = t3) : t3 = $[4];
  let t4;
  $[5] !== t3 ? (t4 = /* @__PURE__ */ jsx(Inline, { space: 2, children: /* @__PURE__ */ jsx(Text, { weight: "medium", children: t3 }) }), $[5] = t3, $[6] = t4) : t4 = $[6];
  let t5;
  $[7] !== t ? (t5 = t("settings.perspectives.description"), $[7] = t, $[8] = t5) : t5 = $[8];
  let t6;
  $[9] !== t5 ? (t6 = /* @__PURE__ */ jsx(Card, { children: /* @__PURE__ */ jsx(Text, { muted: !0, children: t5 }) }), $[9] = t5, $[10] = t6) : t6 = $[10];
  let t7;
  $[11] !== t ? (t7 = t("label.new"), $[11] = t, $[12] = t7) : t7 = $[12];
  let t8;
  $[13] !== t7 ? (t8 = /* @__PURE__ */ jsx(Box, { children: /* @__PURE__ */ jsx(Badge, { tone: "primary", children: t7 }) }), $[13] = t7, $[14] = t8) : t8 = $[14];
  let t9;
  $[15] !== t ? (t9 = /* @__PURE__ */ jsx(Text, { muted: !0, children: /* @__PURE__ */ jsx(Translate, { t, i18nKey: "settings.perspective.preview-drafts-renamed-to-drafts.description" }) }), $[15] = t, $[16] = t9) : t9 = $[16];
  let t10;
  $[17] !== t8 || $[18] !== t9 ? (t10 = /* @__PURE__ */ jsx(Card, { children: /* @__PURE__ */ jsxs(Stack, { space: 2, children: [
    t8,
    t9
  ] }) }), $[17] = t8, $[18] = t9, $[19] = t10) : t10 = $[19];
  let t11;
  $[20] !== t ? (t11 = null, $[20] = t, $[21] = t11) : t11 = $[21];
  let t12;
  $[22] !== t ? (t12 = t("settings.perspectives.action.docs-link"), $[22] = t, $[23] = t12) : t12 = $[23];
  let t13;
  $[24] !== t12 ? (t13 = /* @__PURE__ */ jsx(Card, { children: /* @__PURE__ */ jsx(Text, { children: /* @__PURE__ */ jsxs(PerspectivePopoverLink, { href: "https://www.sanity.io/docs/perspectives", target: "_blank", children: [
    t12,
    " \u2192"
  ] }) }) }), $[24] = t12, $[25] = t13) : t13 = $[25];
  let t14;
  $[26] !== t10 || $[27] !== t11 || $[28] !== t13 || $[29] !== t4 || $[30] !== t6 ? (t14 = /* @__PURE__ */ jsx(PerspectivePopoverContent, { children: /* @__PURE__ */ jsxs(Stack, { space: 4, children: [
    t4,
    t6,
    t10,
    t11,
    t13
  ] }) }), $[26] = t10, $[27] = t11, $[28] = t13, $[29] = t4, $[30] = t6, $[31] = t14) : t14 = $[31];
  let t15;
  $[32] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t15 = /* @__PURE__ */ jsx(Dot, { $tone: "primary" }), $[32] = t15) : t15 = $[32];
  let t16;
  $[33] !== open ? (t16 = /* @__PURE__ */ jsx(Button, { icon: HelpCircleIcon, mode: "bleed", padding: 2, paddingRight: 1, tone: "primary", fontSize: 1, ref: buttonRef, onClick: handleClick, selected: open, children: t15 }), $[33] = open, $[34] = t16) : t16 = $[34];
  let t17;
  return $[35] !== open || $[36] !== t14 || $[37] !== t16 ? (t17 = /* @__PURE__ */ jsx(Popover, { content: t14, placement: "bottom-start", portal: !0, padding: 3, ref: popoverRef, open, children: t16 }), $[35] = open, $[36] = t14, $[37] = t16, $[38] = t17) : t17 = $[38], t17;
}
function _temp$3(o) {
  return !o;
}
const PinnedReleasePerspectiveOption = (t0) => {
  const $ = c(11), {
    pinnedPerspective,
    t
  } = t0, name = typeof pinnedPerspective.selectedPerspective == "object" ? pinnedPerspective.selectedPerspective.metadata.title : pinnedPerspective.selectedPerspectiveName;
  let t1;
  $[0] !== pinnedPerspective || $[1] !== t ? (t1 = hasPinnedPerspective(pinnedPerspective) ? `(${t("settings.perspectives.pinned-release-label")})` : t("settings.perspectives.pinned-release-label"), $[0] = pinnedPerspective, $[1] = t, $[2] = t1) : t1 = $[2];
  const label = t1;
  let t2;
  $[3] !== label || $[4] !== name ? (t2 = [name, label].filter(_temp$2), $[3] = label, $[4] = name, $[5] = t2) : t2 = $[5];
  const text = t2.join(" ");
  let t3;
  $[6] !== pinnedPerspective ? (t3 = hasPinnedPerspective(pinnedPerspective), $[6] = pinnedPerspective, $[7] = t3) : t3 = $[7];
  const t4 = !t3;
  let t5;
  return $[8] !== t4 || $[9] !== text ? (t5 = /* @__PURE__ */ jsx("option", { value: "pinnedRelease", disabled: t4, children: text }), $[8] = t4, $[9] = text, $[10] = t5) : t5 = $[10], t5;
};
function VisionGuiHeader(t0) {
  const $ = c(65), {
    onChangeDataset,
    dataset,
    customApiVersion,
    apiVersion,
    onChangeApiVersion,
    datasets,
    customApiVersionElementRef,
    onCustomApiVersionChange,
    isValidApiVersion,
    onChangePerspective,
    url,
    perspective,
    isScheduledDraftsEnabled
  } = t0, pinnedPerspective = usePerspective(), {
    t
  } = useTranslation(visionLocaleNamespace), operationUrlElement = useRef(null);
  let t1;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t1 = () => {
    const el = operationUrlElement.current;
    if (el)
      try {
        el.select(), document.execCommand("copy");
      } catch {
        console.error("Unable to copy to clipboard :(");
      }
  }, $[0] = t1) : t1 = $[0];
  const handleCopyUrl = t1;
  let t2;
  $[1] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t2 = [1, 4, 8, 12], $[1] = t2) : t2 = $[1];
  let t3;
  $[2] !== t ? (t3 = t("settings.dataset-label"), $[2] = t, $[3] = t3) : t3 = $[3];
  let t4;
  $[4] !== t3 ? (t4 = /* @__PURE__ */ jsx(Card, { paddingTop: 2, paddingBottom: 3, children: /* @__PURE__ */ jsx(StyledLabel, { children: t3 }) }), $[4] = t3, $[5] = t4) : t4 = $[5];
  let t5;
  $[6] !== datasets ? (t5 = datasets.map(_temp2$2), $[6] = datasets, $[7] = t5) : t5 = $[7];
  let t6;
  $[8] !== dataset || $[9] !== onChangeDataset || $[10] !== t5 ? (t6 = /* @__PURE__ */ jsx(Select, { value: dataset, onChange: onChangeDataset, children: t5 }), $[8] = dataset, $[9] = onChangeDataset, $[10] = t5, $[11] = t6) : t6 = $[11];
  let t7;
  $[12] !== t4 || $[13] !== t6 ? (t7 = /* @__PURE__ */ jsx(Box, { padding: 1, column: 2, children: /* @__PURE__ */ jsxs(Stack, { children: [
    t4,
    t6
  ] }) }), $[12] = t4, $[13] = t6, $[14] = t7) : t7 = $[14];
  let t8;
  $[15] !== t ? (t8 = t("settings.api-version-label"), $[15] = t, $[16] = t8) : t8 = $[16];
  let t9;
  $[17] !== t8 ? (t9 = /* @__PURE__ */ jsx(Card, { paddingTop: 2, paddingBottom: 3, children: /* @__PURE__ */ jsx(StyledLabel, { children: t8 }) }), $[17] = t8, $[18] = t9) : t9 = $[18];
  const t10 = customApiVersion === !1 ? apiVersion : "other";
  let t11;
  $[19] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t11 = API_VERSIONS.map(_temp3$1), $[19] = t11) : t11 = $[19];
  let t12;
  $[20] !== t ? (t12 = t("settings.other-api-version-label"), $[20] = t, $[21] = t12) : t12 = $[21];
  let t13;
  $[22] !== t12 ? (t13 = /* @__PURE__ */ jsx("option", { value: "other", children: t12 }, "other"), $[22] = t12, $[23] = t13) : t13 = $[23];
  let t14;
  $[24] !== onChangeApiVersion || $[25] !== t10 || $[26] !== t13 ? (t14 = /* @__PURE__ */ jsxs(Select, { "data-testid": "api-version-selector", value: t10, onChange: onChangeApiVersion, children: [
    t11,
    t13
  ] }), $[24] = onChangeApiVersion, $[25] = t10, $[26] = t13, $[27] = t14) : t14 = $[27];
  let t15;
  $[28] !== t14 || $[29] !== t9 ? (t15 = /* @__PURE__ */ jsx(Box, { padding: 1, column: 2, children: /* @__PURE__ */ jsxs(Stack, { children: [
    t9,
    t14
  ] }) }), $[28] = t14, $[29] = t9, $[30] = t15) : t15 = $[30];
  let t16;
  $[31] !== customApiVersion || $[32] !== customApiVersionElementRef || $[33] !== isValidApiVersion || $[34] !== onCustomApiVersionChange || $[35] !== t ? (t16 = customApiVersion !== !1 && /* @__PURE__ */ jsx(Box, { padding: 1, column: 2, children: /* @__PURE__ */ jsxs(Stack, { children: [
    /* @__PURE__ */ jsx(Card, { paddingTop: 2, paddingBottom: 3, children: /* @__PURE__ */ jsx(StyledLabel, { textOverflow: "ellipsis", children: t("settings.custom-api-version-label") }) }),
    /* @__PURE__ */ jsx(TextInput, { ref: customApiVersionElementRef, value: customApiVersion, onChange: onCustomApiVersionChange, customValidity: isValidApiVersion ? void 0 : t("settings.error.invalid-api-version"), maxLength: 11 })
  ] }) }), $[31] = customApiVersion, $[32] = customApiVersionElementRef, $[33] = isValidApiVersion, $[34] = onCustomApiVersionChange, $[35] = t, $[36] = t16) : t16 = $[36];
  let t17;
  $[37] !== t ? (t17 = t("settings.perspective-label"), $[37] = t, $[38] = t17) : t17 = $[38];
  let t18;
  $[39] !== t17 ? (t18 = /* @__PURE__ */ jsx(Box, { children: /* @__PURE__ */ jsx(StyledLabel, { children: t17 }) }), $[39] = t17, $[40] = t18) : t18 = $[40];
  let t19;
  $[41] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t19 = /* @__PURE__ */ jsx(Box, { children: /* @__PURE__ */ jsx(PerspectivePopover, {}) }), $[41] = t19) : t19 = $[41];
  let t20;
  $[42] !== t18 ? (t20 = /* @__PURE__ */ jsx(Card, { paddingBottom: 1, children: /* @__PURE__ */ jsxs(Inline, { space: 1, children: [
    t18,
    t19
  ] }) }), $[42] = t18, $[43] = t20) : t20 = $[43];
  const t21 = perspective || "default";
  let t22;
  $[44] !== isScheduledDraftsEnabled || $[45] !== pinnedPerspective || $[46] !== t ? (t22 = SUPPORTED_PERSPECTIVES.map((perspectiveName) => perspectiveName === "pinnedRelease" ? /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(PinnedReleasePerspectiveOption, { pinnedPerspective, t }),
    /* @__PURE__ */ jsx("option", { value: "default", children: t("settings.perspectives.default") }, "default"),
    /* @__PURE__ */ jsx("hr", {})
  ] }, "pinnedRelease") : perspectiveName === "scheduledDrafts" ? isScheduledDraftsEnabled ? /* @__PURE__ */ jsx("option", { value: "scheduledDrafts", children: t("settings.perspectives.scheduled-drafts") }, "scheduledDrafts") : null : /* @__PURE__ */ jsx("option", { children: perspectiveName }, perspectiveName)), $[44] = isScheduledDraftsEnabled, $[45] = pinnedPerspective, $[46] = t, $[47] = t22) : t22 = $[47];
  let t23;
  $[48] !== onChangePerspective || $[49] !== t21 || $[50] !== t22 ? (t23 = /* @__PURE__ */ jsx(Select, { value: t21, onChange: onChangePerspective, children: t22 }), $[48] = onChangePerspective, $[49] = t21, $[50] = t22, $[51] = t23) : t23 = $[51];
  let t24;
  $[52] !== t20 || $[53] !== t23 ? (t24 = /* @__PURE__ */ jsx(Box, { padding: 1, column: 2, children: /* @__PURE__ */ jsxs(Stack, { children: [
    t20,
    t23
  ] }) }), $[52] = t20, $[53] = t23, $[54] = t24) : t24 = $[54];
  let t25;
  $[55] !== customApiVersion || $[56] !== t || $[57] !== url ? (t25 = typeof url == "string" ? /* @__PURE__ */ jsx(Box, { padding: 1, flex: 1, column: customApiVersion === !1 ? 6 : 4, children: /* @__PURE__ */ jsxs(Stack, { children: [
    /* @__PURE__ */ jsx(Card, { paddingTop: 2, paddingBottom: 3, children: /* @__PURE__ */ jsxs(StyledLabel, { children: [
      t("query.url"),
      "\xA0",
      /* @__PURE__ */ jsxs(QueryCopyLink, { onClick: handleCopyUrl, children: [
        "[",
        t("action.copy-url-to-clipboard"),
        "]"
      ] })
    ] }) }),
    /* @__PURE__ */ jsxs(Flex, { flex: 1, gap: 1, children: [
      /* @__PURE__ */ jsx(Box, { flex: 1, children: /* @__PURE__ */ jsx(TextInput, { readOnly: !0, type: "url", ref: operationUrlElement, value: url }) }),
      /* @__PURE__ */ jsx(Tooltip, { content: t("action.copy-url-to-clipboard"), children: /* @__PURE__ */ jsx(Button, { "aria-label": t("action.copy-url-to-clipboard"), type: "button", mode: "ghost", icon: CopyIcon, onClick: handleCopyUrl }) })
    ] })
  ] }) }) : /* @__PURE__ */ jsx(Box, { flex: 1 }), $[55] = customApiVersion, $[56] = t, $[57] = url, $[58] = t25) : t25 = $[58];
  let t26;
  return $[59] !== t15 || $[60] !== t16 || $[61] !== t24 || $[62] !== t25 || $[63] !== t7 ? (t26 = /* @__PURE__ */ jsx(Header, { paddingX: 3, paddingY: 2, children: /* @__PURE__ */ jsxs(Grid, { columns: t2, children: [
    t7,
    t15,
    t16,
    t24,
    t25
  ] }) }), $[59] = t15, $[60] = t16, $[61] = t24, $[62] = t25, $[63] = t7, $[64] = t26) : t26 = $[64], t26;
}
function _temp3$1(version) {
  return /* @__PURE__ */ jsx("option", { children: version }, version);
}
function _temp2$2(ds) {
  return /* @__PURE__ */ jsx("option", { children: ds }, ds);
}
function _temp$2(value) {
  return typeof value < "u";
}
function getBlobUrl(content, mimeType) {
  return URL.createObjectURL(new Blob([content], {
    type: mimeType
  }));
}
function getMemoizedBlobUrlResolver(mimeType, stringEncoder) {
  return /* @__PURE__ */ (() => {
    let prevResult = "", prevContent = "";
    return (input) => {
      const content = stringEncoder(input);
      if (!(typeof content != "string" || content === ""))
        return content === prevContent || (prevContent = content, prevResult && URL.revokeObjectURL(prevResult), prevResult = getBlobUrl(content, mimeType)), prevResult;
    };
  })();
}
const getJsonBlobUrl = getMemoizedBlobUrlResolver("application/json", (input) => JSON.stringify(input, null, 2)), getCsvBlobUrl = getMemoizedBlobUrlResolver("text/csv", (input) => json2csv(Array.isArray(input) ? input : [input]).trim()), ErrorCode = styled(Code).withConfig({
  displayName: "ErrorCode",
  componentId: "sc-14lyn64-0"
})`color:${({
  theme
}) => theme.sanity.color.muted.critical.enabled.fg};`;
function QueryErrorDetails(t0) {
  const $ = c(20), {
    error
  } = t0, {
    t
  } = useTranslation(visionLocaleNamespace);
  if (!("details" in error))
    return null;
  const t1 = error.details;
  let T0, details, t2, t3, t4, t5;
  if ($[0] !== error.details) {
    t5 = /* @__PURE__ */ Symbol.for("react.early_return_sentinel");
    bb0: {
      if (details = {
        ...t1,
        ...mapToLegacyDetails(error.details)
      }, !details.line) {
        t5 = null;
        break bb0;
      }
      T0 = ErrorCode, t4 = 1, t2 = details.line, t3 = dashLine(details.column, details.columnEnd);
    }
    $[0] = error.details, $[1] = T0, $[2] = details, $[3] = t2, $[4] = t3, $[5] = t4, $[6] = t5;
  } else
    T0 = $[1], details = $[2], t2 = $[3], t3 = $[4], t4 = $[5], t5 = $[6];
  if (t5 !== /* @__PURE__ */ Symbol.for("react.early_return_sentinel"))
    return t5;
  const t6 = `${t2}
${t3}`;
  let t7;
  $[7] !== T0 || $[8] !== t4 || $[9] !== t6 ? (t7 = /* @__PURE__ */ jsx(T0, { size: t4, children: t6 }), $[7] = T0, $[8] = t4, $[9] = t6, $[10] = t7) : t7 = $[10];
  let t8;
  $[11] !== t ? (t8 = t("query.error.line"), $[11] = t, $[12] = t8) : t8 = $[12];
  const t9 = details.lineNumber;
  let t10;
  $[13] !== t ? (t10 = t("query.error.column"), $[13] = t, $[14] = t10) : t10 = $[14];
  const t11 = `${t8}:   ${t9}
${t10}: ${details.column}`;
  let t12;
  $[15] !== t11 ? (t12 = /* @__PURE__ */ jsx(Box, { marginTop: 4, children: /* @__PURE__ */ jsx(ErrorCode, { size: 1, children: t11 }) }), $[15] = t11, $[16] = t12) : t12 = $[16];
  let t13;
  return $[17] !== t12 || $[18] !== t7 ? (t13 = /* @__PURE__ */ jsxs("div", { children: [
    t7,
    t12
  ] }), $[17] = t12, $[18] = t7, $[19] = t13) : t13 = $[19], t13;
}
function mapToLegacyDetails(details) {
  if (!details || typeof details.query != "string" || typeof details.start != "number")
    return {};
  const {
    query,
    start,
    end
  } = details, lineStart = query.slice(0, start).lastIndexOf(`
`) + 1, lineNumber = (query.slice(0, lineStart).match(/\n/g) || []).length, line = query.slice(lineStart, query.indexOf(`
`, lineStart)), column = start - lineStart, columnEnd = typeof end == "number" ? end - lineStart : void 0;
  return {
    line,
    lineNumber,
    column,
    columnEnd
  };
}
function dashLine(column, columnEnd) {
  const line = "-".repeat(column), hats = "^".repeat(columnEnd ? columnEnd - column : 1);
  return `${line}${hats}`;
}
function QueryErrorDialog(props) {
  const $ = c(7);
  let t0;
  $[0] !== props.error.message ? (t0 = /* @__PURE__ */ jsx(ErrorCode, { size: 1, children: props.error.message }), $[0] = props.error.message, $[1] = t0) : t0 = $[1];
  let t1;
  $[2] !== props.error ? (t1 = /* @__PURE__ */ jsx(QueryErrorDetails, { error: props.error }), $[2] = props.error, $[3] = t1) : t1 = $[3];
  let t2;
  return $[4] !== t0 || $[5] !== t1 ? (t2 = /* @__PURE__ */ jsxs(Stack, { space: 5, marginTop: 2, children: [
    t0,
    t1
  ] }), $[4] = t0, $[5] = t1, $[6] = t2) : t2 = $[6], t2;
}
const ResultViewWrapper = /* @__PURE__ */ styled.div.withConfig({
  displayName: "ResultViewWrapper",
  componentId: "sc-nbqtaq-0"
})(({
  theme
}) => {
  const {
    color,
    fonts,
    space
  } = theme.sanity;
  return css`
    & .json-inspector,
    & .json-inspector .json-inspector__selection {
      font-family: ${fonts.code.family};
      font-size: ${fonts.code.sizes[2].fontSize}px;
      line-height: ${fonts.code.sizes[2].lineHeight}px;
      color: var(--card-code-fg-color);
    }

    & .json-inspector .json-inspector__leaf {
      padding-left: ${rem(space[4])};
    }

    & .json-inspector .json-inspector__leaf.json-inspector__leaf_root {
      padding-top: ${rem(space[0])};
      padding-left: 0;
    }

    & .json-inspector > .json-inspector__leaf_root > .json-inspector__line > .json-inspector__key {
      display: none;
    }

    & .json-inspector .json-inspector__line {
      display: block;
      position: relative;
      cursor: default;
    }

    & .json-inspector .json-inspector__line::after {
      content: '';
      position: absolute;
      top: 0;
      left: -200px;
      right: -50px;
      bottom: 0;
      z-index: -1;
      pointer-events: none;
    }

    & .json-inspector .json-inspector__line:hover::after {
      background: var(--card-code-bg-color);
    }

    & .json-inspector .json-inspector__leaf_composite > .json-inspector__line {
      cursor: pointer;
    }

    & .json-inspector .json-inspector__leaf_composite > .json-inspector__line::before {
      content: '▸ ';
      margin-left: calc(0px - ${rem(space[4])});
      font-size: ${fonts.code.sizes[2].fontSize}px;
      line-height: ${fonts.code.sizes[2].lineHeight}px;
    }

    &
      .json-inspector
      .json-inspector__leaf_expanded.json-inspector__leaf_composite
      > .json-inspector__line::before {
      content: '▾ ';
      font-size: ${fonts.code.sizes[2].fontSize}px;
      line-height: ${fonts.code.sizes[2].lineHeight}px;
    }

    & .json-inspector .json-inspector__radio,
    & .json-inspector .json-inspector__flatpath {
      display: none;
    }

    & .json-inspector .json-inspector__value {
      margin-left: ${rem(space[4] / 2)};
    }

    &
      .json-inspector
      > .json-inspector__leaf_root
      > .json-inspector__line
      > .json-inspector__key
      + .json-inspector__value {
      margin: 0;
    }

    & .json-inspector .json-inspector__key {
      color: ${color.syntax.property};
    }

    & .json-inspector .json-inspector__value_helper,
    & .json-inspector .json-inspector__value_null {
      color: ${color.syntax.constant};
    }

    & .json-inspector .json-inspector__not-found {
      padding-top: ${rem(space[2])};
    }

    & .json-inspector .json-inspector__value_string {
      color: ${color.syntax.string};
      word-break: break-word;
    }

    & .json-inspector .json-inspector__value_boolean {
      color: ${color.syntax.boolean};
    }

    & .json-inspector .json-inspector__value_number {
      color: ${color.syntax.number};
    }

    & .json-inspector .json-inspector__show-original {
      display: inline-block;
      padding: 0 6px;
      cursor: pointer;
    }

    & .json-inspector .json-inspector__show-original:hover {
      color: inherit;
    }

    & .json-inspector .json-inspector__show-original::before {
      content: '↔';
    }

    & .json-inspector .json-inspector__show-original:hover::after {
      content: ' expand';
    }
  `;
}), lru = new LRU({
  maxSize: 5e4
});
function ResultView(props) {
  const $ = c(7), {
    data,
    datasetName
  } = props, workspaceDataset = useDataset();
  if (isRecord(data) || Array.isArray(data)) {
    const t02 = workspaceDataset === datasetName ? DocumentEditLabel : void 0;
    let t12;
    return $[0] !== data || $[1] !== t02 ? (t12 = /* @__PURE__ */ jsx(ResultViewWrapper, { children: /* @__PURE__ */ jsx(JsonInspector, { data, search: !1, isExpanded, onClick: toggleExpanded, interactiveLabel: t02 }) }), $[0] = data, $[1] = t02, $[2] = t12) : t12 = $[2], t12;
  }
  let t0;
  $[3] !== data ? (t0 = JSON.stringify(data), $[3] = data, $[4] = t0) : t0 = $[4];
  let t1;
  return $[5] !== t0 ? (t1 = /* @__PURE__ */ jsx(Code, { language: "json", children: t0 }), $[5] = t0, $[6] = t1) : t1 = $[6], t1;
}
function DocumentEditLabel(props) {
  const $ = c(5);
  if (props.isKey || !props.keypath.endsWith("_id") && !props.keypath.endsWith("_ref"))
    return null;
  let t0;
  $[0] !== props.value ? (t0 = {
    id: props.value
  }, $[0] = props.value, $[1] = t0) : t0 = $[1];
  let t1;
  $[2] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t1 = /* @__PURE__ */ jsx(LinkIcon, {}), $[2] = t1) : t1 = $[2];
  let t2;
  return $[3] !== t0 ? (t2 = /* @__PURE__ */ jsx(IntentLink, { intent: "edit", params: t0, children: t1 }), $[3] = t0, $[4] = t2) : t2 = $[4], t2;
}
function isExpanded(keyPath, value) {
  const cached = lru.get(keyPath);
  if (typeof cached == "boolean")
    return cached;
  const segments = keyPath.split(".", 4);
  return segments.length === 4 ? !1 : Array.isArray(value) ? !0 : isRecord(value) && !segments.some((key) => isArrayKeyOverLimit(key));
}
function toggleExpanded(event) {
  const {
    path
  } = event, current = lru.get(path);
  current !== void 0 && lru.set(path, !current);
}
function isRecord(value) {
  return value !== null && typeof value == "object" && !Array.isArray(value);
}
const numeric = /^\d+$/;
function isArrayKeyOverLimit(segment, limit = 10) {
  return numeric.test(segment) && parseInt(segment, 10) > limit;
}
function preventSave(evt) {
  return evt.preventDefault();
}
function SaveCsvButton(t0) {
  const $ = c(9), {
    blobUrl
  } = t0, {
    t
  } = useTranslation(visionLocaleNamespace), isDisabled = !blobUrl, t1 = isDisabled ? void 0 : "query-result.csv", t2 = isDisabled ? preventSave : void 0;
  let t3;
  $[0] !== blobUrl || $[1] !== isDisabled || $[2] !== t1 || $[3] !== t2 ? (t3 = /* @__PURE__ */ jsx(Button, { as: "a", disabled: isDisabled, download: t1, href: blobUrl, icon: DocumentSheetIcon, mode: "ghost", onClick: t2, text: "CSV", tone: "default" }), $[0] = blobUrl, $[1] = isDisabled, $[2] = t1, $[3] = t2, $[4] = t3) : t3 = $[4];
  const button = t3;
  let t4;
  return $[5] !== button || $[6] !== isDisabled || $[7] !== t ? (t4 = isDisabled ? /* @__PURE__ */ jsx(Tooltip, { content: t("result.save-result-as-csv.not-csv-encodable"), placement: "top", children: button }) : button, $[5] = button, $[6] = isDisabled, $[7] = t, $[8] = t4) : t4 = $[8], t4;
}
function SaveJsonButton(t0) {
  const $ = c(2), {
    blobUrl
  } = t0;
  let t1;
  return $[0] !== blobUrl ? (t1 = /* @__PURE__ */ jsx(Button, { as: "a", download: "query-result.json", href: blobUrl, icon: DocumentSheetIcon, mode: "ghost", text: "JSON", tone: "default" }), $[0] = blobUrl, $[1] = t1) : t1 = $[1], t1;
}
function VisionGuiResult(t0) {
  const $ = c(67), {
    error,
    queryInProgress,
    queryResult,
    listenInProgress,
    listenMutations,
    dataset,
    queryTime,
    e2eTime
  } = t0, {
    t
  } = useTranslation(visionLocaleNamespace), hasResult = !error && !queryInProgress && typeof queryResult < "u";
  let t1;
  $[0] !== hasResult || $[1] !== queryResult ? (t1 = hasResult ? getJsonBlobUrl(queryResult) : "", $[0] = hasResult, $[1] = queryResult, $[2] = t1) : t1 = $[2];
  const jsonUrl = t1;
  let t2;
  $[3] !== hasResult || $[4] !== queryResult ? (t2 = hasResult ? getCsvBlobUrl(queryResult) : "", $[3] = hasResult, $[4] = queryResult, $[5] = t2) : t2 = $[5];
  const csvUrl = t2, t3 = error ? "critical" : "default", t4 = !!error;
  let t5;
  $[6] !== t ? (t5 = t("result.label"), $[6] = t, $[7] = t5) : t5 = $[7];
  let t6;
  $[8] !== t5 ? (t6 = /* @__PURE__ */ jsx(InputBackgroundContainer, { children: /* @__PURE__ */ jsx(Box, { marginLeft: 3, children: /* @__PURE__ */ jsx(StyledLabel, { muted: !0, children: t5 }) }) }), $[8] = t5, $[9] = t6) : t6 = $[9];
  let t7;
  $[10] !== listenInProgress || $[11] !== listenMutations || $[12] !== queryInProgress ? (t7 = (queryInProgress || listenInProgress && listenMutations.length === 0) && /* @__PURE__ */ jsx(Box, { marginTop: 3, children: /* @__PURE__ */ jsx(DelayedSpinner, {}) }), $[10] = listenInProgress, $[11] = listenMutations, $[12] = queryInProgress, $[13] = t7) : t7 = $[13];
  let t8;
  $[14] !== error ? (t8 = error && /* @__PURE__ */ jsx(QueryErrorDialog, { error }), $[14] = error, $[15] = t8) : t8 = $[15];
  let t9;
  $[16] !== dataset || $[17] !== hasResult || $[18] !== queryResult ? (t9 = hasResult && /* @__PURE__ */ jsx(ResultView, { data: queryResult, datasetName: dataset }), $[16] = dataset, $[17] = hasResult, $[18] = queryResult, $[19] = t9) : t9 = $[19];
  let t10;
  $[20] !== dataset || $[21] !== listenInProgress || $[22] !== listenMutations ? (t10 = listenInProgress && listenMutations.length > 0 && /* @__PURE__ */ jsx(ResultView, { data: listenMutations, datasetName: dataset }), $[20] = dataset, $[21] = listenInProgress, $[22] = listenMutations, $[23] = t10) : t10 = $[23];
  let t11;
  $[24] !== t10 || $[25] !== t7 || $[26] !== t8 || $[27] !== t9 ? (t11 = /* @__PURE__ */ jsxs(Box, { padding: 3, paddingTop: 5, children: [
    t7,
    t8,
    t9,
    t10
  ] }), $[24] = t10, $[25] = t7, $[26] = t8, $[27] = t9, $[28] = t11) : t11 = $[28];
  let t12;
  $[29] !== t11 || $[30] !== t6 ? (t12 = /* @__PURE__ */ jsxs(Result, { overflow: "auto", children: [
    t6,
    t11
  ] }), $[29] = t11, $[30] = t6, $[31] = t12) : t12 = $[31];
  let t13;
  $[32] !== t12 || $[33] !== t3 || $[34] !== t4 ? (t13 = /* @__PURE__ */ jsx(ResultInnerContainer, { flex: 1, children: /* @__PURE__ */ jsx(ResultContainer, { flex: 1, overflow: "hidden", tone: t3, $isInvalid: t4, children: t12 }) }), $[32] = t12, $[33] = t3, $[34] = t4, $[35] = t13) : t13 = $[35];
  let t14;
  $[36] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t14 = ["column", "column", "row"], $[36] = t14) : t14 = $[36];
  let t15;
  $[37] !== t ? (t15 = t("result.execution-time-label"), $[37] = t, $[38] = t15) : t15 = $[38];
  let t16;
  $[39] !== queryTime || $[40] !== t ? (t16 = typeof queryTime == "number" ? `${queryTime}ms` : t("result.timing-not-applicable"), $[39] = queryTime, $[40] = t, $[41] = t16) : t16 = $[41];
  let t17;
  $[42] !== t15 || $[43] !== t16 ? (t17 = /* @__PURE__ */ jsx(Box, { children: /* @__PURE__ */ jsxs(Text, { muted: !0, children: [
    t15,
    ":",
    " ",
    t16
  ] }) }), $[42] = t15, $[43] = t16, $[44] = t17) : t17 = $[44];
  let t18;
  $[45] !== t ? (t18 = t("result.end-to-end-time-label"), $[45] = t, $[46] = t18) : t18 = $[46];
  let t19;
  $[47] !== e2eTime || $[48] !== t ? (t19 = typeof e2eTime == "number" ? `${e2eTime}ms` : t("result.timing-not-applicable"), $[47] = e2eTime, $[48] = t, $[49] = t19) : t19 = $[49];
  let t20;
  $[50] !== t18 || $[51] !== t19 ? (t20 = /* @__PURE__ */ jsx(Box, { marginLeft: 4, children: /* @__PURE__ */ jsxs(Text, { muted: !0, children: [
    t18,
    ":",
    " ",
    t19
  ] }) }), $[50] = t18, $[51] = t19, $[52] = t20) : t20 = $[52];
  let t21;
  $[53] !== t17 || $[54] !== t20 ? (t21 = /* @__PURE__ */ jsx(TimingsCard, { paddingX: 4, paddingY: 3, sizing: "border", children: /* @__PURE__ */ jsxs(TimingsTextContainer, { align: "center", children: [
    t17,
    t20
  ] }) }), $[53] = t17, $[54] = t20, $[55] = t21) : t21 = $[55];
  let t22;
  $[56] !== csvUrl || $[57] !== hasResult || $[58] !== jsonUrl || $[59] !== t ? (t22 = hasResult && /* @__PURE__ */ jsx(DownloadsCard, { paddingX: 4, paddingY: 3, sizing: "border", children: /* @__PURE__ */ jsx(SaveResultLabel, { muted: !0, children: /* @__PURE__ */ jsx(Translate, { components: {
    SaveResultButtons: () => /* @__PURE__ */ jsxs(Fragment$1, { children: [
      /* @__PURE__ */ jsx(SaveJsonButton, { blobUrl: jsonUrl }),
      /* @__PURE__ */ jsx(SaveCsvButton, { blobUrl: csvUrl })
    ] })
  }, i18nKey: "result.save-result-as-format", t }) }) }), $[56] = csvUrl, $[57] = hasResult, $[58] = jsonUrl, $[59] = t, $[60] = t22) : t22 = $[60];
  let t23;
  $[61] !== t21 || $[62] !== t22 ? (t23 = /* @__PURE__ */ jsxs(ResultFooter, { justify: "space-between", direction: t14, children: [
    t21,
    t22
  ] }), $[61] = t21, $[62] = t22, $[63] = t23) : t23 = $[63];
  let t24;
  return $[64] !== t13 || $[65] !== t23 ? (t24 = /* @__PURE__ */ jsxs(ResultOuterContainer, { direction: "column", "data-testid": "vision-result", children: [
    t13,
    t23
  ] }), $[64] = t13, $[65] = t23, $[66] = t24) : t24 = $[66], t24;
}
function nodeContains(node, other) {
  return !node || !other ? !1 : node === other || !!(node.compareDocumentPosition(other) & 16);
}
const sanityUrl = /\/(vX|v1|v\d{4}-\d\d-\d\d)\/.*?(?:query|listen)\/(.*?)\?(.*)/, isRunHotkey = (event) => isHotkey("ctrl+enter", event) || isHotkey("mod+enter", event);
function VisionGui(props) {
  const $ = c(210), {
    datasets,
    config,
    projectId,
    defaultDataset
  } = props, toast = useToast(), {
    t
  } = useTranslation(visionLocaleNamespace), {
    perspectiveStack
  } = usePerspective(), isScheduledDraftsEnabled = useScheduledDraftsEnabled(), {
    data: t0
  } = useActiveReleases();
  let t1;
  $[0] !== t0 ? (t1 = t0 === void 0 ? [] : t0, $[0] = t0, $[1] = t1) : t1 = $[1];
  const releases = t1, isDraftModelEnabled = useWorkspace().document.drafts.enabled, editorQueryRef = useRef(null), editorParamsRef = useRef(null), visionRootRef = useRef(null), customApiVersionElementRef = useRef(null), querySubscriptionRef = useRef(void 0), listenSubscriptionRef = useRef(void 0);
  let t2;
  $[2] !== projectId ? (t2 = () => getLocalStorage(projectId || "default"), $[2] = projectId, $[3] = t2) : t2 = $[3];
  const [localStorage2] = useState(t2);
  let t3;
  $[4] !== defaultDataset || $[5] !== localStorage2 ? (t3 = localStorage2.get("dataset", defaultDataset), $[4] = defaultDataset, $[5] = localStorage2, $[6] = t3) : t3 = $[6];
  const storedDataset = t3;
  let t4;
  $[7] !== config.defaultApiVersion || $[8] !== localStorage2 ? (t4 = localStorage2.get("apiVersion", prefixApiVersion(`${config.defaultApiVersion}`)), $[7] = config.defaultApiVersion, $[8] = localStorage2, $[9] = t4) : t4 = $[9];
  const storedApiVersion = t4;
  let t5;
  $[10] !== localStorage2 ? (t5 = localStorage2.get("query", ""), $[10] = localStorage2, $[11] = t5) : t5 = $[11];
  const storedQuery = t5;
  let t6;
  $[12] !== localStorage2 ? (t6 = localStorage2.get("params", `{
  
}`), $[12] = localStorage2, $[13] = t6) : t6 = $[13];
  const storedParams = t6;
  let t7;
  $[14] !== localStorage2 ? (t7 = localStorage2.get("perspective", void 0), $[14] = localStorage2, $[15] = t7) : t7 = $[15];
  const storedPerspective = t7;
  let t8;
  $[16] !== datasets || $[17] !== defaultDataset || $[18] !== storedDataset ? (t8 = () => datasets.includes(storedDataset) ? storedDataset : datasets.includes(defaultDataset) ? defaultDataset : datasets[0], $[16] = datasets, $[17] = defaultDataset, $[18] = storedDataset, $[19] = t8) : t8 = $[19];
  const [dataset, setDataset] = useState(t8);
  let t9;
  $[20] !== storedApiVersion ? (t9 = () => API_VERSIONS.includes(storedApiVersion) ? storedApiVersion : DEFAULT_API_VERSION, $[20] = storedApiVersion, $[21] = t9) : t9 = $[21];
  const [apiVersion, setApiVersion] = useState(t9);
  let t10;
  $[22] !== storedApiVersion ? (t10 = () => API_VERSIONS.includes(storedApiVersion) ? !1 : storedApiVersion, $[22] = storedApiVersion, $[23] = t10) : t10 = $[23];
  const [customApiVersion, setCustomApiVersion] = useState(t10), [perspective, setPerspectiveState] = useState(storedPerspective || "raw");
  let t11;
  $[24] !== customApiVersion ? (t11 = customApiVersion ? validateApiVersion(customApiVersion) : !0, $[24] = customApiVersion, $[25] = t11) : t11 = $[25];
  const isValidApiVersion = t11, [url, setUrl] = useState(void 0);
  let t12;
  $[26] !== storedQuery ? (t12 = () => typeof storedQuery == "string" ? storedQuery : "", $[26] = storedQuery, $[27] = t12) : t12 = $[27];
  const [query, setQuery] = useState(t12), [rawParams, setRawParams] = useState(storedParams);
  let t13;
  $[28] !== rawParams || $[29] !== t ? (t13 = parseParams(rawParams, t), $[28] = rawParams, $[29] = t, $[30] = t13) : t13 = $[30];
  const params = t13, [queryResult, setQueryResult] = useState(void 0);
  let t14;
  $[31] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t14 = [], $[31] = t14) : t14 = $[31];
  const [listenMutations, setListenMutations] = useState(t14), [error, setError] = useState(void 0), [queryTime, setQueryTime] = useState(void 0), [e2eTime, setE2eTime] = useState(void 0), [queryInProgress, setQueryInProgress] = useState(!1), [listenInProgress, setListenInProgress] = useState(!1), [isQueryRecallCollapsed, setIsQueryRecallCollapsed] = useState(!1);
  let t15;
  $[32] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t15 = {
    visionRootRef
  }, $[32] = t15) : t15 = $[32];
  const {
    paneSizeOptions,
    isNarrowBreakpoint
  } = usePaneSize(t15);
  let t16;
  bb0: {
    if (!isScheduledDraftsEnabled) {
      t16 = void 0;
      break bb0;
    }
    let t172;
    if ($[33] !== isDraftModelEnabled || $[34] !== releases) {
      const scheduledDraftReleases = releases.filter(_temp$1), releaseIds = sortReleases(scheduledDraftReleases).map(_temp2$1);
      let t182;
      $[36] !== isDraftModelEnabled ? (t182 = isDraftModelEnabled ? ["drafts"] : ["published"], $[36] = isDraftModelEnabled, $[37] = t182) : t182 = $[37];
      const defaultPerspective = t182;
      t172 = [...releaseIds, ...defaultPerspective], $[33] = isDraftModelEnabled, $[34] = releases, $[35] = t172;
    } else
      t172 = $[35];
    t16 = t172;
  }
  const scheduledDraftsStack = t16, t17 = isValidApiVersion && customApiVersion ? customApiVersion : apiVersion;
  let t18;
  $[38] !== t17 ? (t18 = {
    apiVersion: t17
  }, $[38] = t17, $[39] = t18) : t18 = $[39];
  const _client = useClient(t18);
  let t19;
  $[40] !== _client || $[41] !== apiVersion || $[42] !== customApiVersion || $[43] !== dataset || $[44] !== isValidApiVersion || $[45] !== perspective || $[46] !== perspectiveStack || $[47] !== scheduledDraftsStack ? (t19 = _client.withConfig({
    apiVersion: isValidApiVersion && customApiVersion ? customApiVersion : apiVersion,
    perspective: getActivePerspective({
      visionPerspective: perspective,
      perspectiveStack,
      scheduledDraftsStack
    }),
    dataset,
    allowReconfigure: !0
  }), $[40] = _client, $[41] = apiVersion, $[42] = customApiVersion, $[43] = dataset, $[44] = isValidApiVersion, $[45] = perspective, $[46] = perspectiveStack, $[47] = scheduledDraftsStack, $[48] = t19) : t19 = $[48];
  const client = t19;
  let t20;
  $[49] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t20 = () => {
    querySubscriptionRef.current && (querySubscriptionRef.current.unsubscribe(), querySubscriptionRef.current = void 0);
  }, $[49] = t20) : t20 = $[49];
  const cancelQuerySubscription = t20;
  let t21;
  $[50] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t21 = () => {
    listenSubscriptionRef.current && (listenSubscriptionRef.current.unsubscribe(), listenSubscriptionRef.current = void 0);
  }, $[50] = t21) : t21 = $[50];
  const cancelListenerSubscription = t21;
  let t22;
  $[51] !== apiVersion || $[52] !== client || $[53] !== customApiVersion || $[54] !== dataset || $[55] !== isValidApiVersion || $[56] !== localStorage2 || $[57] !== params.parsed || $[58] !== perspective || $[59] !== perspectiveStack || $[60] !== query || $[61] !== queryInProgress || $[62] !== scheduledDraftsStack || $[63] !== t ? (t22 = (options) => {
    if (queryInProgress) {
      cancelQuerySubscription(), cancelListenerSubscription(), setQueryInProgress(!1);
      return;
    }
    const context = {
      query: options?.query || query,
      dataset: options?.dataset || dataset,
      params: parseParams(JSON.stringify(options?.params || params.parsed, null, 2), t),
      perspective: getActivePerspective({
        visionPerspective: options && "perspective" in options ? options.perspective : perspective,
        perspectiveStack,
        scheduledDraftsStack
      }),
      apiVersion: options?.apiVersion || (customApiVersion && isValidApiVersion ? customApiVersion : apiVersion)
    };
    if (localStorage2.set("query", context.query), localStorage2.set("params", context.params.raw), cancelListenerSubscription(), setQueryInProgress(!context.params.error && !!context.query), setListenInProgress(!1), setListenMutations([]), setError(context.params.error ? new Error(context.params.error) : void 0), setQueryResult(void 0), setQueryTime(void 0), setE2eTime(void 0), context.params.error)
      return;
    const urlQueryOpts = {
      perspective: context.perspective ?? []
    }, ctxClient = client.withConfig({
      apiVersion: context.apiVersion,
      dataset: context.dataset,
      perspective: context.perspective
    }), newUrl = ctxClient.getUrl(ctxClient.getDataUrl("query", encodeQueryString(context.query, context.params.parsed, urlQueryOpts)));
    setUrl(newUrl);
    const queryStart = Date.now();
    querySubscriptionRef.current = ctxClient.observable.fetch(context.query, context.params.parsed, {
      filterResponse: !1,
      tag: "vision"
    }).subscribe({
      next: (res) => {
        setQueryTime(res.ms), setE2eTime(Date.now() - queryStart), setQueryResult(res.result), setQueryInProgress(!1), setError(void 0);
      },
      error: (err) => {
        setError(err), setQueryInProgress(!1);
      }
    });
  }, $[51] = apiVersion, $[52] = client, $[53] = customApiVersion, $[54] = dataset, $[55] = isValidApiVersion, $[56] = localStorage2, $[57] = params.parsed, $[58] = perspective, $[59] = perspectiveStack, $[60] = query, $[61] = queryInProgress, $[62] = scheduledDraftsStack, $[63] = t, $[64] = t22) : t22 = $[64];
  const handleQueryExecution = t22;
  let t23;
  $[65] !== handleQueryExecution || $[66] !== localStorage2 ? (t23 = (newPerspective) => {
    newPerspective !== void 0 && !isSupportedPerspective(newPerspective) || (setPerspectiveState(newPerspective), localStorage2.set("perspective", newPerspective), handleQueryExecution({
      perspective: newPerspective
    }));
  }, $[65] = handleQueryExecution, $[66] = localStorage2, $[67] = t23) : t23 = $[67];
  const setPerspective = t23;
  let t24;
  $[68] !== handleQueryExecution || $[69] !== localStorage2 ? (t24 = (evt) => {
    const newDataset = evt.target.value;
    localStorage2.set("dataset", newDataset), setDataset(newDataset), handleQueryExecution({
      dataset: newDataset
    });
  }, $[68] = handleQueryExecution, $[69] = localStorage2, $[70] = t24) : t24 = $[70];
  const handleChangeDataset = t24;
  let t25;
  $[71] !== handleQueryExecution || $[72] !== localStorage2 ? (t25 = (evt_0) => {
    const newApiVersion = evt_0.target.value;
    if (newApiVersion?.toLowerCase() === "other") {
      setCustomApiVersion("v"), customApiVersionElementRef.current?.focus();
      return;
    }
    setApiVersion(newApiVersion), setCustomApiVersion(!1), localStorage2.set("apiVersion", newApiVersion), handleQueryExecution({
      apiVersion: newApiVersion
    });
  }, $[71] = handleQueryExecution, $[72] = localStorage2, $[73] = t25) : t25 = $[73];
  const handleChangeApiVersion = t25;
  let t26;
  $[74] !== handleQueryExecution || $[75] !== localStorage2 ? (t26 = (evt_1) => {
    const newCustomApiVersion = evt_1.target.value || "";
    setCustomApiVersion(newCustomApiVersion || "v"), validateApiVersion(newCustomApiVersion) && (setApiVersion(newCustomApiVersion), localStorage2.set("apiVersion", newCustomApiVersion), handleQueryExecution({
      apiVersion: newCustomApiVersion
    }));
  }, $[74] = handleQueryExecution, $[75] = localStorage2, $[76] = t26) : t26 = $[76];
  const handleCustomApiVersionChange = t26;
  let t27;
  $[77] !== setPerspective ? (t27 = (evt_2) => {
    const newPerspective_0 = evt_2.target.value;
    setPerspective(newPerspective_0 === "default" ? void 0 : newPerspective_0);
  }, $[77] = setPerspective, $[78] = t27) : t27 = $[78];
  const handleChangePerspective = t27;
  let t28;
  $[79] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t28 = (evt_3) => {
    evt_3.type === "mutation" && setListenMutations((prevMutations) => prevMutations.length === 50 ? [evt_3, ...prevMutations.slice(0, 49)] : [evt_3, ...prevMutations]);
  }, $[79] = t28) : t28 = $[79];
  const handleListenerEvent = t28;
  let t29;
  $[80] !== client || $[81] !== listenInProgress || $[82] !== localStorage2 || $[83] !== params.error || $[84] !== params.parsed || $[85] !== params.raw || $[86] !== query ? (t29 = () => {
    if (listenInProgress) {
      cancelListenerSubscription(), setListenInProgress(!1);
      return;
    }
    const newUrl_0 = client.getDataUrl("listen", encodeQueryString(query, params.parsed, {})), shouldExecute = !params.error && query.trim().length > 0;
    localStorage2.set("query", query), localStorage2.set("params", params.raw), cancelQuerySubscription(), setUrl(newUrl_0), setListenMutations([]), setQueryInProgress(!1), setQueryResult(void 0), setListenInProgress(shouldExecute), setError(params.error ? new Error(params.error) : void 0), setQueryTime(void 0), setE2eTime(void 0), shouldExecute && (listenSubscriptionRef.current = client.listen(query, params.parsed, {
      events: ["mutation", "welcome"],
      includeAllVersions: !0
    }).subscribe({
      next: handleListenerEvent,
      error: (err_0) => {
        setError(err_0), setListenInProgress(!1);
      }
    }));
  }, $[80] = client, $[81] = listenInProgress, $[82] = localStorage2, $[83] = params.error, $[84] = params.parsed, $[85] = params.raw, $[86] = query, $[87] = t29) : t29 = $[87];
  const handleListenExecution = t29;
  let t30;
  $[88] !== localStorage2 ? (t30 = (value) => {
    setRawParams(value), localStorage2.set("params", value);
  }, $[88] = localStorage2, $[89] = t30) : t30 = $[89];
  const handleParamsChange = t30;
  let t31;
  $[90] !== apiVersion || $[91] !== dataset || $[92] !== datasets || $[93] !== perspective || $[94] !== toast ? (t31 = (data) => {
    const match = data.match(sanityUrl);
    if (!match)
      return null;
    const [, usedApiVersion, usedDataset, urlQuery] = match, qs = new URLSearchParams(urlQuery), parts = parseApiQueryString(qs);
    if (!parts)
      return null;
    let newApiVersion_0, newCustomApiVersion_0;
    validateApiVersion(usedApiVersion) && (API_VERSIONS.includes(usedApiVersion) ? (newApiVersion_0 = usedApiVersion, newCustomApiVersion_0 = !1) : newCustomApiVersion_0 = usedApiVersion);
    const newPerspective_1 = isSupportedPerspective(parts.options.perspective) && !isVirtualPerspective(parts.options.perspective) ? parts.options.perspective : void 0;
    return newPerspective_1 && (!isSupportedPerspective(parts.options.perspective) || isVirtualPerspective(parts.options.perspective)) && toast.push({
      closable: !0,
      id: "vision-paste-unsupported-perspective",
      status: "warning",
      title: 'Perspective in pasted url is currently not supported. Falling back to "raw"'
    }), {
      query: parts.query,
      params: parts.params,
      rawParams: JSON.stringify(parts.params, null, 2),
      dataset: datasets.includes(usedDataset) ? usedDataset : dataset,
      apiVersion: newApiVersion_0 || apiVersion,
      customApiVersion: newCustomApiVersion_0,
      perspective: newPerspective_1 || perspective,
      url: data
    };
  }, $[90] = apiVersion, $[91] = dataset, $[92] = datasets, $[93] = perspective, $[94] = toast, $[95] = t31) : t31 = $[95];
  const getStateFromUrl = t31;
  let t32;
  $[96] !== handleQueryExecution || $[97] !== localStorage2 ? (t32 = (parsedUrlObj) => {
    setDataset(parsedUrlObj.dataset), setQuery(parsedUrlObj.query), setRawParams(parsedUrlObj.rawParams), setApiVersion(parsedUrlObj.apiVersion), parsedUrlObj.customApiVersion && setCustomApiVersion(parsedUrlObj.customApiVersion), setPerspectiveState(parsedUrlObj.perspective), setUrl(parsedUrlObj.url), editorQueryRef.current?.resetEditorContent(parsedUrlObj.query), editorParamsRef.current?.resetEditorContent(parsedUrlObj.rawParams), localStorage2.merge({
      query: parsedUrlObj.query,
      params: parsedUrlObj.rawParams,
      dataset: parsedUrlObj.dataset,
      apiVersion: parsedUrlObj.customApiVersion || parsedUrlObj.apiVersion,
      perspective: parsedUrlObj.perspective
    }), handleQueryExecution(parsedUrlObj);
  }, $[96] = handleQueryExecution, $[97] = localStorage2, $[98] = t32) : t32 = $[98];
  const setStateFromParsedUrl = t32;
  let t33;
  $[99] !== getStateFromUrl || $[100] !== setStateFromParsedUrl || $[101] !== toast ? (t33 = (evt_4) => {
    if (!evt_4.clipboardData)
      return;
    const data_0 = evt_4.clipboardData.getData("text/plain");
    evt_4.preventDefault();
    const urlState = getStateFromUrl(data_0);
    urlState && (setStateFromParsedUrl(urlState), toast.push({
      closable: !0,
      id: "vision-paste",
      status: "info",
      title: "Parsed URL to query"
    }));
  }, $[99] = getStateFromUrl, $[100] = setStateFromParsedUrl, $[101] = toast, $[102] = t33) : t33 = $[102];
  const handlePaste = t33;
  let t34;
  $[103] !== handleQueryExecution || $[104] !== params.valid ? (t34 = (event) => {
    const isWithinRoot = visionRootRef.current && nodeContains(visionRootRef.current, event.target);
    isRunHotkey(event) && isWithinRoot && params.valid && (handleQueryExecution(), event.preventDefault(), event.stopPropagation());
  }, $[103] = handleQueryExecution, $[104] = params.valid, $[105] = t34) : t34 = $[105];
  const handleKeyDown = t34;
  let t35, t36;
  $[106] !== handleKeyDown || $[107] !== handlePaste ? (t35 = () => (window.document.addEventListener("paste", handlePaste), window.document.addEventListener("keydown", handleKeyDown), () => {
    window.document.removeEventListener("paste", handlePaste), window.document.removeEventListener("keydown", handleKeyDown);
  }), t36 = [handleKeyDown, handlePaste], $[106] = handleKeyDown, $[107] = handlePaste, $[108] = t35, $[109] = t36) : (t35 = $[108], t36 = $[109]), useEffect(t35, t36);
  let t37, t38;
  $[110] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t37 = () => () => {
    cancelQuerySubscription(), cancelListenerSubscription();
  }, t38 = [cancelQuerySubscription, cancelListenerSubscription], $[110] = t37, $[111] = t38) : (t37 = $[110], t38 = $[111]), useEffect(t37, t38);
  let t39;
  $[112] !== setPerspective ? (t39 = (stack) => {
    stack.length > 0 && setPerspective("pinnedRelease");
  }, $[112] = setPerspective, $[113] = t39) : t39 = $[113];
  const handleStudioPerspectiveChange = useEffectEvent(t39);
  let t40;
  $[114] !== handleStudioPerspectiveChange || $[115] !== perspectiveStack ? (t40 = () => {
    handleStudioPerspectiveChange(perspectiveStack);
  }, $[114] = handleStudioPerspectiveChange, $[115] = perspectiveStack, $[116] = t40) : t40 = $[116];
  let t41;
  $[117] !== perspectiveStack ? (t41 = [perspectiveStack], $[117] = perspectiveStack, $[118] = t41) : t41 = $[118], useEffect(t40, t41);
  let t42;
  $[119] !== client || $[120] !== perspective || $[121] !== perspectiveStack || $[122] !== scheduledDraftsStack ? (t42 = (queryString, queryParams) => {
    const urlQueryOpts_0 = {
      perspective: getActivePerspective({
        visionPerspective: perspective,
        perspectiveStack,
        scheduledDraftsStack
      }) ?? []
    };
    return client.getUrl(client.getDataUrl("query", encodeQueryString(queryString, queryParams, urlQueryOpts_0)));
  }, $[119] = client, $[120] = perspective, $[121] = perspectiveStack, $[122] = scheduledDraftsStack, $[123] = t42) : t42 = $[123];
  const generateUrl = t42;
  let t43;
  $[124] !== apiVersion || $[125] !== customApiVersion || $[126] !== dataset || $[127] !== datasets || $[128] !== handleChangeApiVersion || $[129] !== handleChangeDataset || $[130] !== handleChangePerspective || $[131] !== handleCustomApiVersionChange || $[132] !== isScheduledDraftsEnabled || $[133] !== isValidApiVersion || $[134] !== perspective || $[135] !== url ? (t43 = /* @__PURE__ */ jsx(VisionGuiHeader, { apiVersion, customApiVersion, dataset, datasets, onChangeDataset: handleChangeDataset, onChangeApiVersion: handleChangeApiVersion, customApiVersionElementRef, onCustomApiVersionChange: handleCustomApiVersionChange, isValidApiVersion, onChangePerspective: handleChangePerspective, url, perspective, isScheduledDraftsEnabled }), $[124] = apiVersion, $[125] = customApiVersion, $[126] = dataset, $[127] = datasets, $[128] = handleChangeApiVersion, $[129] = handleChangeDataset, $[130] = handleChangePerspective, $[131] = handleCustomApiVersionChange, $[132] = isScheduledDraftsEnabled, $[133] = isValidApiVersion, $[134] = perspective, $[135] = url, $[136] = t43) : t43 = $[136];
  const t44 = isQueryRecallCollapsed ? window.innerWidth : window.innerWidth - 275, t45 = isNarrowBreakpoint ? "vertical" : "horizontal", t46 = isNarrowBreakpoint ? paneSizeOptions.defaultSize : paneSizeOptions.minSize, t47 = paneSizeOptions.size, t48 = paneSizeOptions.allowResize, t49 = isNarrowBreakpoint ? paneSizeOptions.minSize : 100, t50 = paneSizeOptions.maxSize;
  let t51;
  $[137] !== t ? (t51 = t("query.label"), $[137] = t, $[138] = t51) : t51 = $[138];
  let t52;
  $[139] !== t51 ? (t52 = /* @__PURE__ */ jsx(InputBackgroundContainerLeft, { children: /* @__PURE__ */ jsx(Flex, { children: /* @__PURE__ */ jsx(StyledLabel, { muted: !0, children: t51 }) }) }), $[139] = t51, $[140] = t52) : t52 = $[140];
  let t53;
  $[141] !== query ? (t53 = /* @__PURE__ */ jsx(VisionCodeMirror, { initialValue: query, onChange: setQuery, ref: editorQueryRef }), $[141] = query, $[142] = t53) : t53 = $[142];
  let t54;
  $[143] !== t52 || $[144] !== t53 ? (t54 = /* @__PURE__ */ jsx(InputContainer, { display: "flex", "data-testid": "vision-query-editor", children: /* @__PURE__ */ jsxs(Box, { flex: 1, children: [
    t52,
    t53
  ] }) }), $[143] = t52, $[144] = t53, $[145] = t54) : t54 = $[145];
  let t55;
  $[146] !== handleParamsChange || $[147] !== params.error || $[148] !== params.raw || $[149] !== params.valid ? (t55 = /* @__PURE__ */ jsx(ParamsEditor, { value: params.raw, onChange: handleParamsChange, paramsError: params.error, hasValidParams: params.valid, editorRef: editorParamsRef }), $[146] = handleParamsChange, $[147] = params.error, $[148] = params.raw, $[149] = params.valid, $[150] = t55) : t55 = $[150];
  let t56;
  $[151] !== handleListenExecution || $[152] !== handleQueryExecution || $[153] !== listenInProgress || $[154] !== params.valid || $[155] !== queryInProgress ? (t56 = /* @__PURE__ */ jsx(VisionGuiControls, { hasValidParams: params.valid, queryInProgress, listenInProgress, onQueryExecution: handleQueryExecution, onListenExecution: handleListenExecution }), $[151] = handleListenExecution, $[152] = handleQueryExecution, $[153] = listenInProgress, $[154] = params.valid, $[155] = queryInProgress, $[156] = t56) : t56 = $[156];
  let t57;
  $[157] !== t55 || $[158] !== t56 ? (t57 = /* @__PURE__ */ jsxs(InputContainer, { display: "flex", children: [
    t55,
    t56
  ] }), $[157] = t55, $[158] = t56, $[159] = t57) : t57 = $[159];
  let t58;
  $[160] !== paneSizeOptions.allowResize || $[161] !== paneSizeOptions.maxSize || $[162] !== paneSizeOptions.size || $[163] !== t46 || $[164] !== t49 || $[165] !== t54 || $[166] !== t57 ? (t58 = /* @__PURE__ */ jsx(Box, { height: "stretch", flex: 1, children: /* @__PURE__ */ jsxs(SplitPane, { className: "sidebarPanes", split: "horizontal", defaultSize: t46, size: t47, allowResize: t48, minSize: t49, maxSize: t50, primary: "first", children: [
    t54,
    t57
  ] }) }), $[160] = paneSizeOptions.allowResize, $[161] = paneSizeOptions.maxSize, $[162] = paneSizeOptions.size, $[163] = t46, $[164] = t49, $[165] = t54, $[166] = t57, $[167] = t58) : t58 = $[167];
  let t59;
  $[168] !== dataset || $[169] !== e2eTime || $[170] !== error || $[171] !== listenInProgress || $[172] !== listenMutations || $[173] !== queryInProgress || $[174] !== queryResult || $[175] !== queryTime ? (t59 = /* @__PURE__ */ jsx(VisionGuiResult, { error, queryInProgress, queryResult, listenInProgress, listenMutations, dataset, queryTime, e2eTime }), $[168] = dataset, $[169] = e2eTime, $[170] = error, $[171] = listenInProgress, $[172] = listenMutations, $[173] = queryInProgress, $[174] = queryResult, $[175] = queryTime, $[176] = t59) : t59 = $[176];
  let t60;
  $[177] !== t45 || $[178] !== t58 || $[179] !== t59 ? (t60 = /* @__PURE__ */ jsx(Box, { height: "stretch", flex: 1, children: /* @__PURE__ */ jsxs(SplitPane, { className: "sidebarPanes", split: t45, minSize: 300, children: [
    t58,
    t59
  ] }) }), $[177] = t45, $[178] = t58, $[179] = t59, $[180] = t60) : t60 = $[180];
  let t61;
  $[181] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t61 = {
    position: "relative",
    height: "100%"
  }, $[181] = t61) : t61 = $[181];
  let t62;
  $[182] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t62 = {
    position: "absolute",
    left: -32,
    top: "50%",
    transform: "translateY(-50%)",
    zIndex: 100,
    pointerEvents: "auto"
  }, $[182] = t62) : t62 = $[182];
  let t63;
  $[183] !== isQueryRecallCollapsed ? (t63 = () => setIsQueryRecallCollapsed(!isQueryRecallCollapsed), $[183] = isQueryRecallCollapsed, $[184] = t63) : t63 = $[184];
  let t64;
  $[185] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t64 = {
    display: "flex",
    alignItems: "center",
    height: "100%"
  }, $[185] = t64) : t64 = $[185];
  let t65;
  $[186] !== isQueryRecallCollapsed ? (t65 = /* @__PURE__ */ jsx("div", { style: t64, children: isQueryRecallCollapsed ? /* @__PURE__ */ jsx(ChevronLeftIcon, {}) : /* @__PURE__ */ jsx(ChevronRightIcon, {}) }), $[186] = isQueryRecallCollapsed, $[187] = t65) : t65 = $[187];
  let t66;
  $[188] !== t63 || $[189] !== t65 ? (t66 = /* @__PURE__ */ jsx(Button, { mode: "ghost", padding: 2, style: t62, onClick: t63, children: t65 }), $[188] = t63, $[189] = t65, $[190] = t66) : t66 = $[190];
  let t67;
  $[191] !== params.parsed ? (t67 = params.parsed || {}, $[191] = params.parsed, $[192] = t67) : t67 = $[192];
  let t68;
  $[193] !== generateUrl || $[194] !== getStateFromUrl || $[195] !== query || $[196] !== setStateFromParsedUrl || $[197] !== t67 || $[198] !== url ? (t68 = /* @__PURE__ */ jsx(QueryRecall, { url, getStateFromUrl, setStateFromParsedUrl, currentQuery: query, currentParams: t67, generateUrl }), $[193] = generateUrl, $[194] = getStateFromUrl, $[195] = query, $[196] = setStateFromParsedUrl, $[197] = t67, $[198] = url, $[199] = t68) : t68 = $[199];
  let t69;
  $[200] !== t66 || $[201] !== t68 ? (t69 = /* @__PURE__ */ jsxs(Box, { style: t61, children: [
    t66,
    t68
  ] }), $[200] = t66, $[201] = t68, $[202] = t69) : t69 = $[202];
  let t70;
  $[203] !== t44 || $[204] !== t60 || $[205] !== t69 ? (t70 = /* @__PURE__ */ jsx(SplitpaneContainer, { flex: "auto", children: /* @__PURE__ */ jsxs(SplitPane, { minSize: 800, defaultSize: window.innerWidth - 275, size: t44, maxSize: -225, primary: "first", children: [
    t60,
    t69
  ] }) }), $[203] = t44, $[204] = t60, $[205] = t69, $[206] = t70) : t70 = $[206];
  let t71;
  return $[207] !== t43 || $[208] !== t70 ? (t71 = /* @__PURE__ */ jsxs(Root, { direction: "column", height: "fill", ref: visionRootRef, sizing: "border", overflow: "hidden", "data-testid": "vision-root", children: [
    t43,
    t70
  ] }), $[207] = t43, $[208] = t70, $[209] = t71) : t71 = $[209], t71;
}
function _temp2$1(release_0) {
  return getReleaseIdFromReleaseDocumentId(release_0._id);
}
function _temp$1(release) {
  return isCardinalityOneRelease(release) && release.state === "scheduled";
}
function useDatasets(t0) {
  const $ = c(7), {
    client,
    datasets: configDatasets
  } = t0;
  let t1;
  bb0: {
    if (Array.isArray(configDatasets)) {
      let t22;
      $[0] !== configDatasets ? (t22 = of(configDatasets), $[0] = configDatasets, $[1] = t22) : t22 = $[1], t1 = t22;
      break bb0;
    }
    let t2;
    if ($[2] !== client.observable.datasets || $[3] !== configDatasets) {
      let t3;
      $[5] !== configDatasets ? (t3 = (result) => typeof configDatasets == "function" ? configDatasets(result).map(_temp) : result.map(_temp2), $[5] = configDatasets, $[6] = t3) : t3 = $[6], t2 = client.observable.datasets.list().pipe(map$1(t3), catchError(_temp3)), $[2] = client.observable.datasets, $[3] = configDatasets, $[4] = t2;
    } else
      t2 = $[4];
    t1 = t2;
  }
  return useObservable(t1, null);
}
function _temp3(err) {
  return of(err);
}
function _temp2(ds) {
  return ds.name;
}
function _temp(d) {
  return d.name;
}
function VisionContainer(props) {
  const $ = c(19);
  let t0;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t0 = {
    apiVersion: "v2025-06-27"
  }, $[0] = t0) : t0 = $[0];
  const datasetsClient = useClient(t0);
  let t1;
  $[1] !== datasetsClient || $[2] !== props.config.datasets ? (t1 = {
    client: datasetsClient,
    datasets: props.config.datasets
  }, $[1] = datasetsClient, $[2] = props.config.datasets, $[3] = t1) : t1 = $[3];
  const loadedDatasets = useDatasets(t1);
  if (!loadedDatasets) {
    let t22;
    return $[4] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t22 = /* @__PURE__ */ jsx(Flex, { align: "center", height: "fill", justify: "center", children: /* @__PURE__ */ jsx(DelayedSpinner, {}) }), $[4] = t22) : t22 = $[4], t22;
  }
  let t2;
  $[5] !== loadedDatasets || $[6] !== props.client ? (t2 = loadedDatasets instanceof Error ? [props.client.config().dataset || "production"] : loadedDatasets, $[5] = loadedDatasets, $[6] = props.client, $[7] = t2) : t2 = $[7];
  const datasets = t2;
  let t3;
  $[8] !== props.client ? (t3 = props.client.config(), $[8] = props.client, $[9] = t3) : t3 = $[9];
  const projectId = t3.projectId;
  let t4;
  $[10] !== datasets || $[11] !== props.client || $[12] !== props.config.defaultDataset ? (t4 = props.config.defaultDataset || props.client.config().dataset || datasets[0], $[10] = datasets, $[11] = props.client, $[12] = props.config.defaultDataset, $[13] = t4) : t4 = $[13];
  const defaultDataset = t4;
  let t5;
  return $[14] !== datasets || $[15] !== defaultDataset || $[16] !== projectId || $[17] !== props ? (t5 = /* @__PURE__ */ jsx(VisionGui, { ...props, datasets, projectId, defaultDataset }, projectId), $[14] = datasets, $[15] = defaultDataset, $[16] = projectId, $[17] = props, $[18] = t5) : t5 = $[18], t5;
}
class VisionErrorBoundary extends Component {
  constructor(props) {
    super(props), this.state = {
      error: null,
      numRetries: 0
    };
  }
  static getDerivedStateFromError(error) {
    return {
      error: error instanceof Error ? error.message : `${error}`
    };
  }
  handleRetryRender = () => this.setState((prev) => ({
    error: null,
    numRetries: prev.numRetries + 1
  }));
  handleRetryWithCacheClear = () => {
    clearLocalStorage(), this.handleRetryRender();
  };
  render() {
    if (!this.state.error)
      return this.props.children;
    const message = this.state.error, withCacheClear = this.state.numRetries > 0;
    return /* @__PURE__ */ jsx(Card, { height: "fill", overflow: "auto", paddingY: [4, 5, 6, 7], paddingX: 4, sizing: "border", tone: "critical", children: /* @__PURE__ */ jsx(Container, { width: 3, children: /* @__PURE__ */ jsxs(Stack, { space: 4, children: [
      /* @__PURE__ */ jsx("div", { children: /* @__PURE__ */ jsx(Button, { onClick: withCacheClear ? this.handleRetryWithCacheClear : this.handleRetryRender, text: withCacheClear ? "Clear cache and retry" : "Retry", tone: "default" }) }),
      /* @__PURE__ */ jsx(Heading, { children: "An error occurred" }),
      /* @__PURE__ */ jsx(Card, { border: !0, radius: 2, overflow: "auto", padding: 4, tone: "inherit", children: /* @__PURE__ */ jsx(Stack, { space: 4, children: message && /* @__PURE__ */ jsx(Code, { size: 1, children: /* @__PURE__ */ jsxs("strong", { children: [
        "Error: ",
        message
      ] }) }) }) })
    ] }) }) });
  }
}
function SanityVision(props) {
  const $ = c(6);
  let t0;
  $[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel") ? (t0 = {
    apiVersion: "1"
  }, $[0] = t0) : t0 = $[0];
  const client = useClient(t0);
  let t1;
  $[1] !== props.tool.options ? (t1 = {
    defaultApiVersion: DEFAULT_API_VERSION,
    ...props.tool.options
  }, $[1] = props.tool.options, $[2] = t1) : t1 = $[2];
  const config = t1;
  let t2;
  return $[3] !== client || $[4] !== config ? (t2 = /* @__PURE__ */ jsx(VisionErrorBoundary, { children: /* @__PURE__ */ jsx(VisionContainer, { client, config }) }), $[3] = client, $[4] = config, $[5] = t2) : t2 = $[5], t2;
}
export {
  SanityVision as default
};
