// @sanity/cli/dist/actions/manifest/referenceTransformer.js
import { ensureCustomTitle } from './transformerUtils.js';
/**
 * Transforms a reference schema type to its manifest representation
 */ export function transformReference(reference, retainCustomTypeProps) {
    return {
        to: (reference.to ?? []).map((type)=>{
            return {
                ...retainCustomTypeProps(type),
                type: type.name
            };
        })
    };
}
/**
 * Transforms a cross-dataset reference schema type to its manifest representation
 */ export function transformCrossDatasetReference(reference) {
    return {
        to: (reference.to ?? []).map((crossDataset)=>{
            const preview = crossDataset.preview?.select ? {
                preview: {
                    select: crossDataset.preview.select
                }
            } : {};
            return {
                type: crossDataset.type,
                ...ensureCustomTitle(crossDataset.type, crossDataset.title),
                ...preview
            };
        })
    };
}
/**
 * Transforms a global document reference schema type to its manifest representation
 */ export function transformGlobalDocumentReference(reference) {
    return {
        to: (reference.to ?? []).map((crossDataset)=>{
            const preview = crossDataset.preview?.select ? {
                preview: {
                    select: crossDataset.preview.select
                }
            } : {};
            return {
                type: crossDataset.type,
                ...ensureCustomTitle(crossDataset.type, crossDataset.title),
                ...preview
            };
        })
    };
}
