// @sanity/cli/dist/server/devServer.js
import { createServer } from 'vite';
import { extendViteConfigWithUserConfig, getViteConfig } from '../actions/build/getViteConfig.js';
import { writeSanityRuntime } from '../actions/build/writeSanityRuntime.js';
import { serverDebug } from './serverDebug.js';
const debug = serverDebug.extend('dev');
export async function startDevServer(options) {
    const { basePath, cwd, entry, httpHost, httpPort, isApp, reactCompiler, reactStrictMode, schemaExtraction, typegen, vite: extendViteConfig } = options;
    debug('Writing Sanity runtime files');
    const watcher = await writeSanityRuntime({
        basePath,
        cwd,
        entry,
        isApp,
        reactStrictMode,
        watch: true
    });
    debug('Resolving vite config');
    const mode = 'development';
    let viteConfig = await getViteConfig({
        basePath,
        cwd,
        isApp,
        mode: 'development',
        reactCompiler,
        schemaExtraction,
        server: {
            host: httpHost,
            port: httpPort
        },
        typegen
    });
    // Extend Vite configuration with user-provided config
    if (extendViteConfig) {
        viteConfig = await extendViteConfigWithUserConfig({
            command: 'serve',
            mode
        }, viteConfig, extendViteConfig);
    }
    debug('Creating vite server');
    const server = await createServer(viteConfig);
    debug('Listening on specified port');
    await server.listen();
    return {
        close: async ()=>{
            if (watcher) {
                await watcher.close();
            }
            await server.close();
        },
        server,
        watcher
    };
}
