// @sanity/cli/dist/util/parseArguments.js
import minimist from 'minimist';
/**
 * Parse the arguments from the command line
 *
 * @param argv - The arguments from the command line
 * @returns The parsed arguments
 */ export function parseArguments(argv = process.argv) {
    const args = argv.slice(2);
    const { '--': extraArguments, _, version, ...extOptions } = minimist(args, {
        '--': true,
        boolean: [
            'version'
        ],
        string: [
            '_'
        ]
    });
    const [groupOrCommand, ...argsWithoutOptions] = _;
    const finalExtraArguments = [
        ...extraArguments || [],
        ...argv.filter((arg)=>arg.startsWith('-'))
    ];
    // oclif allows to run `sanity help` or `sanity help <command>`
    // It does not fire the hooks on `--help` so this is okay to track
    const hasHelp = args.includes('help');
    // We only have global debug via env var
    const hasDebug = !!process.env.DEBUG;
    return {
        groupOrCommand,
        argsWithoutOptions,
        argv,
        extOptions,
        extraArguments: finalExtraArguments,
        coreOptions: {
            debug: hasDebug,
            help: hasHelp,
            version
        }
    };
}
