// @sanity/cli/dist/util/sharedFlags.js
import { Flags } from '@oclif/core';
const OVERRIDE_SUFFIX = ' (overrides CLI configuration)';
/**
 * Returns a `--project-id` / `-p` flag definition.
 *
 * Locked: flag name (`project-id`), char (`p`), `helpValue` (`<id>`), and parse (trims + validates non-empty).
 */ export function getProjectIdFlag(options) {
    const { description: baseDescription, helpGroup, semantics, ...rest } = options;
    const isOverride = semantics === 'override';
    const description = (baseDescription ?? 'Project ID to use') + (isOverride ? OVERRIDE_SUFFIX : '');
    return {
        'project-id': Flags.string({
            description,
            helpGroup: helpGroup ?? (isOverride ? 'OVERRIDE' : undefined),
            helpValue: '<id>',
            ...rest,
            char: 'p',
            parse: async (input)=>{
                const trimmed = input.trim();
                if (trimmed === '') {
                    throw new Error('`--project-id` cannot be empty if provided');
                }
                return trimmed;
            }
        })
    };
}
/**
 * Returns a `--dataset` / `-d` flag definition.
 *
 * Locked: flag name (`dataset`), char (`d`), `helpValue` (`<name>`), and parse (trims + validates non-empty).
 */ export function getDatasetFlag(options) {
    const { description: baseDescription, helpGroup, semantics, ...rest } = options;
    const isOverride = semantics === 'override';
    const description = (baseDescription ?? 'Dataset to use') + (isOverride ? OVERRIDE_SUFFIX : '');
    return {
        dataset: Flags.string({
            description,
            helpGroup: helpGroup ?? (isOverride ? 'OVERRIDE' : undefined),
            helpValue: '<name>',
            ...rest,
            char: 'd',
            parse: async (input)=>{
                const trimmed = input.trim();
                if (trimmed === '') {
                    throw new Error('`--dataset` cannot be empty if provided');
                }
                return trimmed;
            }
        })
    };
}
